<img src="https://habrastorage.org/webt/ia/m9/zk/iam9zkyzqebnf_okxipihkgjwnw.jpeg" />
    
**<center>[mlcourse.ai](https://mlcourse.ai) – Open Machine Learning Course** </center><br>
Author: [Yury Kashnitskiy](https://yorko.github.io) (@yorko). [mlcourse.ai](https://mlcourse.ai) is powered by [OpenDataScience (ods.ai)](https://ods.ai/) © 2017—2021

## <center>Assignment #6. Task</center><a class="tocSkip">
### <center> Beating benchmarks in "How good is your Medium article?"</center><a class="tocSkip">
    
[Competition](https://www.kaggle.com/c/how-good-is-your-medium-article). The task is to beat "Assignment 6 baseline" (~1.45 Public LB score). You can refer to [this simple Ridge baseline](https://www.kaggle.com/kashnitsky/ridge-countvectorizer-baseline?rvi=1).

*For discussions, please stick to [ODS Slack](https://opendatascience.slack.com/), channel __#mlcourse_ai_eng__, pinned thread __#a6_bonus__. If you are sure that something is not 100% correct, please leave your feedback there*

-----

# Imports

In [1]:
# !pip install xgboost nltk dill optuna

In [1]:
import json
import os
import pickle
from collections import defaultdict
from functools import partial
from html.parser import HTMLParser
from pathlib import Path

import dill as pickle
import numpy as np
import optuna
import pandas as pd
import xgboost as xgb
from nltk import TweetTokenizer
from scipy.sparse import csr_matrix, hstack
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS, TfidfVectorizer
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from tqdm.notebook import tqdm

# Params

In [2]:
SEED = 42
PATH_TO_DATA = "data"
PATH_TO_SAVE_DIR = "prepared_data"

# Cleaning HTML

The following code will help to throw away all HTML tags from an article content.

In [3]:
from html.parser import HTMLParser


class MLStripper(HTMLParser):
    def __init__(self):
        self.reset()
        self.strict = False
        self.convert_charrefs = True
        self.fed = []

    def handle_data(self, d):
        self.fed.append(d)

    def get_data(self):
        return "".join(self.fed)


def strip_tags(html):
    s = MLStripper()
    s.feed(html)
    return s.get_data()

# Reading JSON

Supplementary function to read a JSON line without crashing on escape characters.

In [4]:
def read_json_line(line=None):
    result = None
    try:
        result = json.loads(line)
    except Exception as e:
        # Find the offending character index:
        idx_to_replace = int(str(e).split(" ")[-1].replace(")", ""))
        # Remove the offending character:
        new_line = list(line)
        new_line[idx_to_replace] = " "
        new_line = "".join(new_line)
        return read_json_line(line=new_line)
    return result

Extract features `content`, `published`, `title` and `author`, write them to separate files for train and test sets.

In [5]:
def check_file(filename):
    file = Path(filename)
    return file.is_file()

In [6]:
def read_from_disk(features, path_to_save, prefix):
    reload_files = False
    d = defaultdict()
    for feature in features:
        filename = os.path.join(path_to_save, prefix + "_" + feature + ".txt")
        if check_file(filename):
            print(f"Reading from disk {filename}")
            with open(filename, "rb") as fp:
                d[feature] = pickle.load(fp)
        else:
            reload_files = True

    return reload_files, d

In [7]:
def extract_features_and_write(path_to_data, path_to_save, inp_filename, is_train=True):
    titles = []
    contents = []
    dates = []

    authors = []
    features = ["content", "published", "title", "author"]
    prefix = "train" if is_train else "test"

    Path(path_to_save).mkdir(parents=True, exist_ok=True)

    reload_files, d = read_from_disk(features, path_to_save, prefix)

    if reload_files:

        feature_files = [
            open(os.path.join(path_to_save, "{}_{}.txt".format(prefix, feat)), "w", encoding="utf-8")
            for feat in features
        ]

        with open(os.path.join(path_to_data, inp_filename), encoding="utf-8") as inp_json_file:

            for line in tqdm(inp_json_file, desc=f"Reading {prefix} json files"):
                json_data = read_json_line(line)

                title = json_data["title"].replace("\n", " ").replace("\t", " ").replace("\r", " ").replace("\xa0", " ")
                content = strip_tags(
                    json_data["content"].replace("\n", " ").replace("\t", " ").replace("\r", " ").replace("\xa0", " ")
                )
                published = json_data["published"]
                author = json_data["meta_tags"]["author"]
                authors_name = json_data["meta_tags"]["author"]

                titles.append(title)
                contents.append(content)
                dates.append(published)
                authors.append(authors_name)

        d = {"content": contents, "published": dates, "title": titles, "author": authors}
        for feature in features:
            filename = prefix + "_" + feature + ".txt"
            with open(os.path.join(path_to_save, filename), "wb") as fp:
                pickle.dump(d[feature], fp)
    else:
        titles, contents, dates, authors = d["title"], d["content"], d["published"], d["author"]

    return titles, contents, dates, authors

Download the [competition data](https://www.kaggle.com/c/how-good-is-your-medium-article/data) and place it where it's convenient for you. You can modify the path to data below.

In [8]:
train_titles, train_contents, train_dates, train_authors = extract_features_and_write(
    PATH_TO_DATA, PATH_TO_SAVE_DIR, "train.json", is_train=True
)
test_titles, test_contents, test_dates, test_authors = extract_features_and_write(
    PATH_TO_DATA, PATH_TO_SAVE_DIR, "test.json", is_train=False
)

Reading from disk prepared_data\train_content.txt
Reading from disk prepared_data\train_published.txt
Reading from disk prepared_data\train_title.txt
Reading from disk prepared_data\train_author.txt
Reading from disk prepared_data\test_content.txt
Reading from disk prepared_data\test_published.txt
Reading from disk prepared_data\test_title.txt
Reading from disk prepared_data\test_author.txt


# Feature engineering

**Add the following groups of features:**
* Tf-Idf with article content:
  * ngram_range=(1, 2)
  * max_features=100000
* Tf-Idf with article titles:
  * ngram_range=(1, 2)
  * max_features=100000
* Time features: 
  * publication hour, 
  * time of the day 
  * weekend or not
* Bag of authors  
i.e. One-Hot-Encoded author names

In [9]:
TITLE_NGRAMS = (1, 2)  # for tf-idf on titles
CONTENT_NGRAMS = (1, 2)  # for tf-idf on contents
MAX_FEATURES = 100_000  # for tf-idf

MEAN_TEST_TARGET = 4.33328  # what we got by submitting all zeros

RIDGE_WEIGHT = 0.6  # weight of Ridge predictions in a blend with XGBoost

## Tokenizer

In [10]:
tokenizer = TweetTokenizer()
assert tokenizer.tokenize("Now I'm a man") == ["Now", "I'm", "a", "man"]

In [11]:
def tokenize(text):
    return tokenizer.tokenize(text)

## Doing TF-IDF vectorization for articles

In [12]:
art_vec_params = {
    "ngram_range": CONTENT_NGRAMS,
    "max_features": MAX_FEATURES,
    "tokenizer": tokenize,
    "stop_words": ENGLISH_STOP_WORDS,
}

In [13]:
article_vectorizer_params_str = f"{art_vec_params['ngram_range']}_{art_vec_params['max_features']}"
article_vec_name = f"vectorizer_article_{article_vectorizer_params_str}.pickle"

In [14]:
%%time
article_vec_is_saved = check_file(article_vec_name)
if article_vec_is_saved:
    vectorizer_article = pickle.load(open(article_vec_name, "rb"))
else:
    vectorizer_article = TfidfVectorizer(**article_vectorizer_params)
    vectorizer_article.fit(train_contents)
    pickle.dump(vectorizer_article, open(article_vec_name, "wb"))

X_train_article = vectorizer_article.transform(train_contents)
X_test_article = vectorizer_article.transform(test_contents)

CPU times: total: 12min 25s
Wall time: 12min 26s


## Doing TF-IDF vectorization for titles

In [15]:
title_vec_params = {
    "ngram_range": TITLE_NGRAMS,
    "max_features": MAX_FEATURES,
    "tokenizer": tokenize,
    "stop_words": ENGLISH_STOP_WORDS,
}

In [16]:
title_vectorizer_params_str = f"{title_vec_params['ngram_range']}_{title_vec_params['max_features']}"
title_vec_name = f"vectorizer_title_{title_vectorizer_params_str}.pickle"

In [17]:
%%time
title_vec_is_saved = check_file(title_vec_name)
if title_vec_is_saved:
    vectorizer_title = pickle.load(open(title_vec_name, "rb"))
else:
    vectorizer_title = TfidfVectorizer(**title_vec_params)
    vectorizer_title.fit(train_titles)
    pickle.dump(vectorizer_title, open(title_vec_name, "wb"))

X_train_title = vectorizer_title.transform(train_titles)
X_test_title = vectorizer_title.transform(test_titles)

CPU times: total: 6.27 s
Wall time: 6.27 s


## Preparing time features

In [18]:
# Time features
def add_time_features(dates):
    scaler = StandardScaler()
    hour = scaler.fit_transform(np.array([date.hour for date in dates]).reshape(-1, 1))
    weekday = scaler.fit_transform(np.array([date.weekday() for date in dates]).reshape(-1, 1))
    morning = scaler.fit_transform(((hour >= 7) & (hour <= 11)).astype("int").reshape(-1, 1))
    day = scaler.fit_transform(((hour >= 12) & (hour <= 18)).astype("int").reshape(-1, 1))
    evening = scaler.fit_transform(((hour >= 19) & (hour <= 23)).astype("int").reshape(-1, 1))
    night = scaler.fit_transform(((hour >= 0) & (hour <= 6)).astype("int").reshape(-1, 1))
    weekend_temp = np.array([date.weekday() for date in dates]).reshape(-1, 1)
    weekend = scaler.fit_transform(((weekend_temp >= 5) & (weekend_temp <= 6)).astype("int").reshape(-1, 1))

    feature_names = ["morning", "day", "evening", "night", "weekday"]
    time_features = pd.DataFrame(
        list(zip(morning.flatten(), day.flatten(), evening.flatten(), night.flatten(), weekend.flatten())),
        columns=feature_names,
    )
    sparse_time_features = csr_matrix(time_features.values)
    return sparse_time_features, feature_names

In [19]:
train_times = pd.to_datetime([date["$date"] for date in train_dates])
test_times = pd.to_datetime([date["$date"] for date in test_dates])

X_train_time_features_sparse, time_feature_names = add_time_features(train_times)
X_test_time_features_sparse, _ = add_time_features(test_times)

## Doing bag of authors

In [20]:
authors = np.unique(train_authors + test_authors)
enc = OneHotEncoder(handle_unknown="ignore")
enc.fit(authors.reshape(-1, 1))
enc.categories_
X_train_author_sparse = enc.transform(np.array(train_authors).reshape(-1, 1)).toarray()
X_test_author_sparse = enc.transform(np.array(test_authors).reshape(-1, 1)).toarray()

## Preparing additional features

In [21]:
train_len = [len(article) for article in train_contents]
test_len = [len(article) for article in test_contents]
scaler = StandardScaler()

X_train_len_sparse = scaler.fit_transform(np.array(train_len).reshape(-1, 1))
X_test_len_sparse = scaler.fit_transform(np.array(test_len).reshape(-1, 1))

## Joining features

**Join all sparse matrices.**

In [22]:
X_train_sparse = hstack(
    [X_train_article, X_train_title, X_train_author_sparse, X_train_time_features_sparse, X_train_len_sparse]
).tocsr()

X_test_sparse = hstack(
    [X_test_article, X_test_title, X_test_author_sparse, X_test_time_features_sparse, X_test_len_sparse]
).tocsr()

### Saving dataset for training

In [29]:
pickle.dump(X_trapickle, open("X_train_sparse.pickle", "wb"))
pickle.dump(X_test_sparse, open("X_test_sparse.pickle", "wb"))
pickle.dump(y_train, open("y_train.pickle", "wb"))

### Reading dataset from disk

In [8]:
X_train_sparse = pickle.load(open("X_train_sparse.pickle", "rb"))
X_test_sparse = pickle.load(open("X_test_sparse.pickle", "rb"))
y_train = pickle.load(open("y_train.pickle", "rb"))

# Read train target and split data for validation

In [23]:
train_target = pd.read_csv(os.path.join(PATH_TO_DATA, "train_log1p_recommends.csv"), index_col="id")
y_train = train_target["log_recommends"].values

In [24]:
train_part_size = int(0.7 * train_target.shape[0])

X_train_part_sparse = X_train_sparse[:train_part_size, :]
y_train_part = y_train[:train_part_size]

X_valid_sparse = X_train_sparse[train_part_size:, :]
y_valid = y_train[train_part_size:]

# Training

## Ridge

### Hyperparameters tuning

In [66]:
# alpha_values = np.logspace(-2, 2, 40)
alpha_values = np.logspace(0, 0.1, 40)

logit_grid_searcher = GridSearchCV(
    estimator=Ridge(random_state=SEED),
    param_grid={"alpha": alpha_values},
    scoring="neg_mean_absolute_error",
    n_jobs=6,
    cv=3,
    verbose=3,
)

logit_grid_searcher.fit(X_train_sparse, y_train)

print(logit_grid_searcher.best_params_)

Fitting 3 folds for each of 40 candidates, totalling 120 fits
{'alpha': 1.1867492730446552}


In [77]:
logit_grid_searcher.best_score_

-1.1032014144706075

{'alpha': 1.1867492730446552}

### Training with best hyperparameters

In [67]:
ridge = Ridge(random_state=SEED, alpha=1.1867492730446552)
ridge.fit(X_train_sparse, y_train)
ridge_test_pred = ridge.predict(X_test_sparse)

## XGBoost

https://www.kaggle.com/code/hamzaghanmi/xgboost-catboost-using-optuna?scriptVersionId=94510532  
https://blog.cambridgespark.com/hyperparameter-tuning-in-xgboost-4ff9100a3b2f  
https://www.youtube.com/watch?v=5nYqK-HaoKY  
https://www.kaggle.com/code/prashant111/a-guide-on-xgboost-hyperparameters-tuning/notebook#2.-XGBoost-hyperparameters-  

### Tuning # 1

In [25]:
def optimize(trial, X, y):
    train_x, valid_x, train_y, valid_y = train_test_split(X, y, test_size=0.20, random_state=SEED)

    dtrain = xgb.DMatrix(train_x, label=train_y)
    dvalid = xgb.DMatrix(valid_x, label=valid_y)

    param = {
        "seed": SEED,
        "verbosity": 1,
        "objective": "reg:pseudohubererror",
        "eval_metric": "mae",
        # gpu_hist for faster training
        "tree_method": "gpu_hist",
        # defines booster, gblinear for linear functions.
        "booster": trial.suggest_categorical("booster", ["gbtree", "dart"]),
        # defines how selective algorithm is.
        "grow_policy": trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"]),
        # learning rate should be tuned also
        "eta": trial.suggest_float("eta", 1e-8, 0.3, log=True),
        # the maximum depth of a tree
        "max_depth": trial.suggest_int("max_depth", 3, 6),
        # L2 regularization weight.
        "lambda": trial.suggest_float("lambda", 1e-8, 1.0, log=True),
        # L1 regularization weight.
        "alpha": trial.suggest_float("alpha", 1e-8, 1.0, log=True),
        # sampling ratio for training data.
        "subsample": trial.suggest_float("subsample", 0.5, 1.0),
        # sampling according to each tree.
        "colsample_bytree": trial.suggest_float("colsample_bytree", 0.2, 1.0),
        # sampling for each split.
        "colsample_bylevel": trial.suggest_float("colsample_bylevel", 0.2, 1.0),
        # minimum child weight, larger the term more conservative the tree.
        "min_child_weight": trial.suggest_int("min_child_weight", 2, 10),
        # defines how selective algorithm is.
        "gamma": trial.suggest_float("gamma", 1e-8, 1.0, log=True),
    }

    if param["booster"] == "dart":
        param["sample_type"] = trial.suggest_categorical("sample_type", ["uniform", "weighted"])
        param["normalize_type"] = trial.suggest_categorical("normalize_type", ["tree", "forest"])
        param["rate_drop"] = trial.suggest_float("rate_drop", 1e-8, 1.0, log=True)
        param["skip_drop"] = trial.suggest_float("skip_drop", 1e-8, 1.0, log=True)

    bst = xgb.train(param, dtrain, num_boost_round=10, evals=[(dtrain, 'train') , (dvalid, 'valid')])
    preds = bst.predict(dvalid)
    mae = mean_absolute_error(valid_y, preds)
    return mae

In [26]:
study = optuna.create_study(direction="minimize")
optimization_function = partial(optimize, X=X_train_sparse, y=y_train)
study.optimize(optimization_function, n_trials=100, timeout=600)

print("Number of finished trials: ", len(study.trials))
print("Best trial:")

trial = study.best_trial

print("  Value: {}".format(trial.value))
print("  Params: ")

for key, value in trial.params.items():
    print("    {}: {}".format(key, value))

[I 2022-05-14 19:38:56,659] A new study created in memory with name: no-name-886ec20b-2bf4-40bd-ae03-b555c85eb88d


[0]	train-mae:2.55233	valid-mae:2.54836
[1]	train-mae:2.55233	valid-mae:2.54836
[2]	train-mae:2.55233	valid-mae:2.54836
[3]	train-mae:2.55232	valid-mae:2.54835
[4]	train-mae:2.55232	valid-mae:2.54835
[5]	train-mae:2.55232	valid-mae:2.54835
[6]	train-mae:2.55232	valid-mae:2.54835
[7]	train-mae:2.55231	valid-mae:2.54835
[8]	train-mae:2.55231	valid-mae:2.54834
[9]	train-mae:2.55231	valid-mae:2.54834


[I 2022-05-14 19:39:04,808] Trial 0 finished with value: 2.5483408523720095 and parameters: {'booster': 'dart', 'grow_policy': 'lossguide', 'eta': 3.7770108294162213e-07, 'max_depth': 6, 'lambda': 0.0005710434639914917, 'alpha': 1.153364178878217e-06, 'subsample': 0.8749220448797093, 'colsample_bytree': 0.27517746843506635, 'min_child_weight': 5, 'gamma': 7.66891214645706e-08, 'sample_type': 'uniform', 'normalize_type': 'tree', 'rate_drop': 5.015019784891533e-07, 'skip_drop': 2.2897671762452735e-07}. Best is trial 0 with value: 2.5483408523720095.


[0]	train-mae:2.55233	valid-mae:2.54836
[1]	train-mae:2.55233	valid-mae:2.54836
[2]	train-mae:2.55233	valid-mae:2.54836
[3]	train-mae:2.55233	valid-mae:2.54836
[4]	train-mae:2.55233	valid-mae:2.54836
[5]	train-mae:2.55233	valid-mae:2.54836
[6]	train-mae:2.55233	valid-mae:2.54836
[7]	train-mae:2.55233	valid-mae:2.54836
[8]	train-mae:2.55233	valid-mae:2.54836
[9]	train-mae:2.55233	valid-mae:2.54836


[I 2022-05-14 19:39:20,790] Trial 1 finished with value: 2.5483608178954347 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 5.17482669985481e-08, 'max_depth': 6, 'lambda': 2.218577651417594e-06, 'alpha': 0.11590049305730092, 'subsample': 0.6432499359449777, 'colsample_bytree': 0.5378731314746252, 'min_child_weight': 8, 'gamma': 2.196056696121911e-05, 'sample_type': 'uniform', 'normalize_type': 'tree', 'rate_drop': 0.00427605742956224, 'skip_drop': 0.07688597486456346}. Best is trial 0 with value: 2.5483408523720095.


[0]	train-mae:2.51574	valid-mae:2.51165
[1]	train-mae:2.47923	valid-mae:2.47535
[2]	train-mae:2.44310	valid-mae:2.43935
[3]	train-mae:2.40085	valid-mae:2.39642
[4]	train-mae:2.36604	valid-mae:2.36140
[5]	train-mae:2.33476	valid-mae:2.33001
[6]	train-mae:2.30586	valid-mae:2.30145
[7]	train-mae:2.27987	valid-mae:2.27559
[8]	train-mae:2.25532	valid-mae:2.25112
[9]	train-mae:2.23282	valid-mae:2.22864


[I 2022-05-14 19:39:27,570] Trial 2 finished with value: 2.228638204776627 and parameters: {'booster': 'gbtree', 'grow_policy': 'depthwise', 'eta': 0.008749221689992048, 'max_depth': 5, 'lambda': 8.067930210384114e-06, 'alpha': 1.4046267495827775e-06, 'subsample': 0.7184652292086466, 'colsample_bytree': 0.32068265270744095, 'min_child_weight': 8, 'gamma': 4.3585110379234395e-06}. Best is trial 2 with value: 2.228638204776627.


[0]	train-mae:2.55230	valid-mae:2.54833
[1]	train-mae:2.55226	valid-mae:2.54829
[2]	train-mae:2.55222	valid-mae:2.54825
[3]	train-mae:2.55218	valid-mae:2.54821
[4]	train-mae:2.55214	valid-mae:2.54817
[5]	train-mae:2.55210	valid-mae:2.54813
[6]	train-mae:2.55207	valid-mae:2.54809
[7]	train-mae:2.55203	valid-mae:2.54805
[8]	train-mae:2.55199	valid-mae:2.54802
[9]	train-mae:2.55195	valid-mae:2.54798


[I 2022-05-14 19:39:34,269] Trial 3 finished with value: 2.5479764526638973 and parameters: {'booster': 'gbtree', 'grow_policy': 'lossguide', 'eta': 6.062587670435519e-06, 'max_depth': 5, 'lambda': 0.013271218195993459, 'alpha': 2.028277614073638e-06, 'subsample': 0.5360273850637675, 'colsample_bytree': 0.3980946234740246, 'min_child_weight': 3, 'gamma': 0.000295240255401177}. Best is trial 2 with value: 2.228638204776627.


[0]	train-mae:2.34338	valid-mae:2.33691
[1]	train-mae:2.19812	valid-mae:2.19157
[2]	train-mae:2.10304	valid-mae:2.09654
[3]	train-mae:2.00632	valid-mae:1.99926
[4]	train-mae:1.93536	valid-mae:1.92846
[5]	train-mae:1.87563	valid-mae:1.86951
[6]	train-mae:1.82202	valid-mae:1.81673
[7]	train-mae:1.77506	valid-mae:1.77021
[8]	train-mae:1.72702	valid-mae:1.72369
[9]	train-mae:1.69101	valid-mae:1.68842


[I 2022-05-14 19:39:41,309] Trial 4 finished with value: 1.6884237326394544 and parameters: {'booster': 'dart', 'grow_policy': 'lossguide', 'eta': 0.048621051957583726, 'max_depth': 3, 'lambda': 0.01074665690690664, 'alpha': 0.006527345263568742, 'subsample': 0.6826846787307952, 'colsample_bytree': 0.663853477393143, 'min_child_weight': 10, 'gamma': 0.2557208570681174, 'sample_type': 'weighted', 'normalize_type': 'tree', 'rate_drop': 0.05742727462833867, 'skip_drop': 2.5379084438812864e-05}. Best is trial 4 with value: 1.6884237326394544.


[0]	train-mae:2.55163	valid-mae:2.54765
[1]	train-mae:2.55091	valid-mae:2.54693
[2]	train-mae:2.55022	valid-mae:2.54623
[3]	train-mae:2.54952	valid-mae:2.54551
[4]	train-mae:2.54881	valid-mae:2.54480
[5]	train-mae:2.54812	valid-mae:2.54409
[6]	train-mae:2.54741	valid-mae:2.54338
[7]	train-mae:2.54672	valid-mae:2.54268
[8]	train-mae:2.54603	valid-mae:2.54199
[9]	train-mae:2.54535	valid-mae:2.54129


[I 2022-05-14 19:39:49,981] Trial 5 finished with value: 2.541294092176935 and parameters: {'booster': 'gbtree', 'grow_policy': 'depthwise', 'eta': 0.00011159378486595666, 'max_depth': 5, 'lambda': 0.05718660101912864, 'alpha': 0.00042190641555143245, 'subsample': 0.9409484434497386, 'colsample_bytree': 0.45614979165486247, 'min_child_weight': 4, 'gamma': 6.524373596207272e-05}. Best is trial 4 with value: 1.6884237326394544.


[0]	train-mae:2.55233	valid-mae:2.54836
[1]	train-mae:2.55233	valid-mae:2.54836
[2]	train-mae:2.55233	valid-mae:2.54836
[3]	train-mae:2.55233	valid-mae:2.54836
[4]	train-mae:2.55233	valid-mae:2.54836
[5]	train-mae:2.55233	valid-mae:2.54836
[6]	train-mae:2.55233	valid-mae:2.54836
[7]	train-mae:2.55233	valid-mae:2.54836
[8]	train-mae:2.55233	valid-mae:2.54836
[9]	train-mae:2.55233	valid-mae:2.54836


[I 2022-05-14 19:39:58,132] Trial 6 finished with value: 2.5483618899947063 and parameters: {'booster': 'gbtree', 'grow_policy': 'lossguide', 'eta': 2.8774112404857327e-08, 'max_depth': 3, 'lambda': 0.0004965919877884663, 'alpha': 0.00013162711703224692, 'subsample': 0.9300576163511796, 'colsample_bytree': 0.8460829942768895, 'min_child_weight': 7, 'gamma': 2.5633251377126083e-08}. Best is trial 4 with value: 1.6884237326394544.


[0]	train-mae:2.54980	valid-mae:2.54584
[1]	train-mae:2.54723	valid-mae:2.54330
[2]	train-mae:2.54442	valid-mae:2.54047
[3]	train-mae:2.54138	valid-mae:2.53739
[4]	train-mae:2.53820	valid-mae:2.53422
[5]	train-mae:2.53552	valid-mae:2.53154
[6]	train-mae:2.53230	valid-mae:2.52830
[7]	train-mae:2.52920	valid-mae:2.52518
[8]	train-mae:2.52553	valid-mae:2.52146
[9]	train-mae:2.52196	valid-mae:2.51786


[I 2022-05-14 19:40:01,644] Trial 7 finished with value: 2.517857581342391 and parameters: {'booster': 'gbtree', 'grow_policy': 'lossguide', 'eta': 0.0006589199311822324, 'max_depth': 3, 'lambda': 8.057827434144808e-08, 'alpha': 1.9708546013780077e-08, 'subsample': 0.6160194712499356, 'colsample_bytree': 0.34008838461634283, 'min_child_weight': 5, 'gamma': 0.006306138226039154}. Best is trial 4 with value: 1.6884237326394544.


[0]	train-mae:1.92282	valid-mae:1.92539
[1]	train-mae:1.84117	valid-mae:1.85400
[2]	train-mae:1.62010	valid-mae:1.63595
[3]	train-mae:1.55785	valid-mae:1.57646
[4]	train-mae:1.51748	valid-mae:1.54235
[5]	train-mae:1.49257	valid-mae:1.51683
[6]	train-mae:1.47157	valid-mae:1.50056
[7]	train-mae:1.45785	valid-mae:1.48943
[8]	train-mae:1.44704	valid-mae:1.47714
[9]	train-mae:1.43672	valid-mae:1.46796


[I 2022-05-14 19:40:09,327] Trial 8 finished with value: 1.4679629695072574 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.2304234358613151, 'max_depth': 4, 'lambda': 0.002673934696130189, 'alpha': 1.8653682927641606e-07, 'subsample': 0.8821092456378741, 'colsample_bytree': 0.47603014187895365, 'min_child_weight': 9, 'gamma': 0.4205963171983195, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 0.01150126975249651, 'skip_drop': 0.016071903405917284}. Best is trial 8 with value: 1.4679629695072574.


[0]	train-mae:2.55233	valid-mae:2.54836
[1]	train-mae:2.55232	valid-mae:2.54835
[2]	train-mae:2.55232	valid-mae:2.54835
[3]	train-mae:2.55231	valid-mae:2.54834
[4]	train-mae:2.55231	valid-mae:2.54834
[5]	train-mae:2.55230	valid-mae:2.54833
[6]	train-mae:2.55230	valid-mae:2.54833
[7]	train-mae:2.55229	valid-mae:2.54832
[8]	train-mae:2.55229	valid-mae:2.54832
[9]	train-mae:2.55228	valid-mae:2.54831


[I 2022-05-14 19:40:15,270] Trial 9 finished with value: 2.548313479310738 and parameters: {'booster': 'dart', 'grow_policy': 'lossguide', 'eta': 1.0525370182513726e-06, 'max_depth': 6, 'lambda': 0.038582814265544225, 'alpha': 0.00038923173140541375, 'subsample': 0.577173937579589, 'colsample_bytree': 0.20306404567339087, 'min_child_weight': 6, 'gamma': 2.144102495810526e-05, 'sample_type': 'weighted', 'normalize_type': 'tree', 'rate_drop': 0.0006240941866228517, 'skip_drop': 4.124796600605576e-07}. Best is trial 8 with value: 1.4679629695072574.


[0]	train-mae:1.99110	valid-mae:1.99191
[1]	train-mae:1.77969	valid-mae:1.78727
[2]	train-mae:1.63447	valid-mae:1.64018
[3]	train-mae:1.54855	valid-mae:1.55819
[4]	train-mae:1.49169	valid-mae:1.50328
[5]	train-mae:1.45309	valid-mae:1.46575
[6]	train-mae:1.42380	valid-mae:1.43839
[7]	train-mae:1.40484	valid-mae:1.41771
[8]	train-mae:1.39074	valid-mae:1.40405
[9]	train-mae:1.37819	valid-mae:1.39315


[I 2022-05-14 19:40:27,615] Trial 10 finished with value: 1.3931484339194795 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.1777221613181068, 'max_depth': 4, 'lambda': 0.735059953337419, 'alpha': 2.065590301856722e-08, 'subsample': 0.8243745305884371, 'colsample_bytree': 0.7213144174966618, 'min_child_weight': 10, 'gamma': 0.40451455030067945, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 2.514583326167942e-06, 'skip_drop': 0.28440523034667514}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.32005	valid-mae:2.31647
[1]	train-mae:2.18551	valid-mae:2.18349
[2]	train-mae:2.08564	valid-mae:2.08398
[3]	train-mae:1.99913	valid-mae:1.99935
[4]	train-mae:1.92566	valid-mae:1.92675
[5]	train-mae:1.86339	valid-mae:1.86535
[6]	train-mae:1.80808	valid-mae:1.81054
[7]	train-mae:1.75915	valid-mae:1.76203
[8]	train-mae:1.71635	valid-mae:1.72065
[9]	train-mae:1.67782	valid-mae:1.68270


[I 2022-05-14 19:40:42,405] Trial 11 finished with value: 1.6826971821547745 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.048592000823844995, 'max_depth': 4, 'lambda': 0.47293598270484083, 'alpha': 1.116027544914122e-08, 'subsample': 0.8194266393686171, 'colsample_bytree': 0.7515985715254349, 'min_child_weight': 10, 'gamma': 0.9084500941458908, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 6.982499232973514e-07, 'skip_drop': 0.7797560901657195}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.07871	valid-mae:2.07146
[1]	train-mae:1.80480	valid-mae:1.80166
[2]	train-mae:1.67293	valid-mae:1.67350
[3]	train-mae:1.58636	valid-mae:1.58994
[4]	train-mae:1.52337	valid-mae:1.52810
[5]	train-mae:1.48029	valid-mae:1.48460
[6]	train-mae:1.44888	valid-mae:1.45344
[7]	train-mae:1.42508	valid-mae:1.43051
[8]	train-mae:1.40539	valid-mae:1.41084
[9]	train-mae:1.39098	valid-mae:1.39799


[I 2022-05-14 19:41:03,167] Trial 12 finished with value: 1.3979932365881071 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.15494080988145445, 'max_depth': 4, 'lambda': 0.2808681095736559, 'alpha': 6.824116221521478e-08, 'subsample': 0.8019797066648477, 'colsample_bytree': 0.989486100011082, 'min_child_weight': 9, 'gamma': 0.01940972353526939, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 1.966524672787188e-05, 'skip_drop': 0.005505578549154645}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.54324	valid-mae:2.53928
[1]	train-mae:2.53414	valid-mae:2.53007
[2]	train-mae:2.52502	valid-mae:2.52087
[3]	train-mae:2.51596	valid-mae:2.51171
[4]	train-mae:2.50757	valid-mae:2.50328
[5]	train-mae:2.49894	valid-mae:2.49453
[6]	train-mae:2.49061	valid-mae:2.48618
[7]	train-mae:2.48264	valid-mae:2.47815
[8]	train-mae:2.47388	valid-mae:2.46924
[9]	train-mae:2.46519	valid-mae:2.46046


[I 2022-05-14 19:41:19,871] Trial 13 finished with value: 2.460459738061768 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.0018978508719152438, 'max_depth': 4, 'lambda': 0.9813842713223127, 'alpha': 9.89865884481778e-08, 'subsample': 0.7817438834366534, 'colsample_bytree': 0.9407381116172092, 'min_child_weight': 10, 'gamma': 0.012868154455516654, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 1.5635159208596866e-05, 'skip_drop': 0.0016281387893071063}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.07860	valid-mae:2.08448
[1]	train-mae:1.65102	valid-mae:1.64764
[2]	train-mae:1.56174	valid-mae:1.56426
[3]	train-mae:1.50978	valid-mae:1.51414
[4]	train-mae:1.47958	valid-mae:1.48324
[5]	train-mae:1.45876	valid-mae:1.46390
[6]	train-mae:1.44442	valid-mae:1.45223
[7]	train-mae:1.42601	valid-mae:1.43391
[8]	train-mae:1.41080	valid-mae:1.41987
[9]	train-mae:1.39984	valid-mae:1.41005


[I 2022-05-14 19:41:42,449] Trial 14 finished with value: 1.4100454639468238 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.28560140676554974, 'max_depth': 4, 'lambda': 0.3145143063807252, 'alpha': 1.4453778123299537e-05, 'subsample': 0.80648852078838, 'colsample_bytree': 0.9775319570652365, 'min_child_weight': 8, 'gamma': 0.008947789905007358, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 2.0720571070728683e-08, 'skip_drop': 0.0008321747088565688}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.51972	valid-mae:2.51551
[1]	train-mae:2.48862	valid-mae:2.48427
[2]	train-mae:2.45587	valid-mae:2.45109
[3]	train-mae:2.42847	valid-mae:2.42328
[4]	train-mae:2.40285	valid-mae:2.39765
[5]	train-mae:2.37865	valid-mae:2.37343
[6]	train-mae:2.35574	valid-mae:2.35032
[7]	train-mae:2.33409	valid-mae:2.32879
[8]	train-mae:2.31326	valid-mae:2.30804
[9]	train-mae:2.29384	valid-mae:2.28849


[I 2022-05-14 19:41:54,662] Trial 15 finished with value: 2.2884931473973076 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.006586779646686553, 'max_depth': 4, 'lambda': 4.625824253289441e-05, 'alpha': 1.9953042515157604e-07, 'subsample': 0.7448522542213505, 'colsample_bytree': 0.6958679091222479, 'min_child_weight': 9, 'gamma': 0.046790307251470105, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 2.632456273734129e-05, 'skip_drop': 0.5832115991255364}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.55149	valid-mae:2.54753
[1]	train-mae:2.55067	valid-mae:2.54671
[2]	train-mae:2.54985	valid-mae:2.54588
[3]	train-mae:2.54903	valid-mae:2.54506
[4]	train-mae:2.54822	valid-mae:2.54424
[5]	train-mae:2.54741	valid-mae:2.54343
[6]	train-mae:2.54659	valid-mae:2.54261
[7]	train-mae:2.54579	valid-mae:2.54182
[8]	train-mae:2.54499	valid-mae:2.54102
[9]	train-mae:2.54420	valid-mae:2.54022


[I 2022-05-14 19:42:14,399] Trial 16 finished with value: 2.5402216713475045 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.00011343261459693063, 'max_depth': 5, 'lambda': 2.772603335639107e-08, 'alpha': 1.2313314034183e-05, 'subsample': 0.861617352898818, 'colsample_bytree': 0.830398750278658, 'min_child_weight': 2, 'gamma': 0.000780662214917002, 'sample_type': 'weighted', 'normalize_type': 'forest', 'rate_drop': 3.7237419862058693e-06, 'skip_drop': 3.361738247796573e-05}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.38186	valid-mae:2.37642
[1]	train-mae:2.27566	valid-mae:2.27109
[2]	train-mae:2.19452	valid-mae:2.19060
[3]	train-mae:2.12768	valid-mae:2.12426
[4]	train-mae:2.06609	valid-mae:2.06298
[5]	train-mae:2.00846	valid-mae:2.00639
[6]	train-mae:1.95680	valid-mae:1.95520
[7]	train-mae:1.91103	valid-mae:1.90940
[8]	train-mae:1.87003	valid-mae:1.86890
[9]	train-mae:1.83264	valid-mae:1.83160


[I 2022-05-14 19:42:25,217] Trial 17 finished with value: 1.8316047921351382 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.03408572884098304, 'max_depth': 3, 'lambda': 0.08529853329102745, 'alpha': 0.6038932368930172, 'subsample': 0.9654464401447346, 'colsample_bytree': 0.8445625781514395, 'min_child_weight': 9, 'gamma': 0.08498349488957688, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 2.262207237147564e-08, 'skip_drop': 0.024240059324880617}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.54897	valid-mae:2.54497
[1]	train-mae:2.54587	valid-mae:2.54187
[2]	train-mae:2.54300	valid-mae:2.53897
[3]	train-mae:2.54009	valid-mae:2.53608
[4]	train-mae:2.53714	valid-mae:2.53311
[5]	train-mae:2.53418	valid-mae:2.53012
[6]	train-mae:2.53110	valid-mae:2.52702
[7]	train-mae:2.52806	valid-mae:2.52392
[8]	train-mae:2.52511	valid-mae:2.52094
[9]	train-mae:2.52215	valid-mae:2.51798


[I 2022-05-14 19:42:33,808] Trial 18 finished with value: 2.5179802190444316 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.0006168965307956796, 'max_depth': 4, 'lambda': 0.0037832521336102817, 'alpha': 1.1483308979302459e-08, 'subsample': 0.7015192849345733, 'colsample_bytree': 0.5977475955341676, 'min_child_weight': 7, 'gamma': 0.0009988440949696559, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 0.5971355624617724, 'skip_drop': 0.0012492020150926355}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.55224	valid-mae:2.54827
[1]	train-mae:2.55215	valid-mae:2.54818
[2]	train-mae:2.55206	valid-mae:2.54809
[3]	train-mae:2.55197	valid-mae:2.54800
[4]	train-mae:2.55188	valid-mae:2.54791
[5]	train-mae:2.55179	valid-mae:2.54782
[6]	train-mae:2.55170	valid-mae:2.54773
[7]	train-mae:2.55161	valid-mae:2.54764
[8]	train-mae:2.55152	valid-mae:2.54755
[9]	train-mae:2.55144	valid-mae:2.54746


[I 2022-05-14 19:43:04,367] Trial 19 finished with value: 2.547459554063221 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 1.652563117508126e-05, 'max_depth': 5, 'lambda': 0.14344354841472515, 'alpha': 1.5688691248542465e-05, 'subsample': 0.7665996186982847, 'colsample_bytree': 0.9970917108090258, 'min_child_weight': 7, 'gamma': 4.576521625150711e-07, 'sample_type': 'weighted', 'normalize_type': 'forest', 'rate_drop': 0.00013555414321562105, 'skip_drop': 0.0958671078461802}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.51131	valid-mae:2.50723
[1]	train-mae:2.47216	valid-mae:2.46768
[2]	train-mae:2.43968	valid-mae:2.43484
[3]	train-mae:2.40999	valid-mae:2.40508
[4]	train-mae:2.38129	valid-mae:2.37602
[5]	train-mae:2.35499	valid-mae:2.34965
[6]	train-mae:2.32992	valid-mae:2.32452
[7]	train-mae:2.30639	valid-mae:2.30117
[8]	train-mae:2.28386	valid-mae:2.27887
[9]	train-mae:2.26327	valid-mae:2.25859


[I 2022-05-14 19:43:13,593] Trial 20 finished with value: 2.2585902324010343 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.00795145008862173, 'max_depth': 3, 'lambda': 6.310828906306785e-07, 'alpha': 9.248397163001559e-08, 'subsample': 0.8356085319835227, 'colsample_bytree': 0.7614806115431795, 'min_child_weight': 9, 'gamma': 0.034514793323943804, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 4.89220338477111e-07, 'skip_drop': 2.7393838731849783e-06}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.05309	valid-mae:2.05438
[1]	train-mae:1.66383	valid-mae:1.66434
[2]	train-mae:1.55367	valid-mae:1.55602
[3]	train-mae:1.49496	valid-mae:1.50288
[4]	train-mae:1.45992	valid-mae:1.46917
[5]	train-mae:1.43472	valid-mae:1.44309
[6]	train-mae:1.41864	valid-mae:1.42832
[7]	train-mae:1.40681	valid-mae:1.41761
[8]	train-mae:1.39361	valid-mae:1.40947
[9]	train-mae:1.38409	valid-mae:1.40028


[I 2022-05-14 19:43:33,427] Trial 21 finished with value: 1.4002753472753648 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.26702264014190114, 'max_depth': 4, 'lambda': 0.9486252328795519, 'alpha': 1.4440825565645808e-05, 'subsample': 0.7971468857909836, 'colsample_bytree': 0.9290427026836316, 'min_child_weight': 8, 'gamma': 0.004432209172236841, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 1.1777619250419617e-08, 'skip_drop': 0.0011924481328482014}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:1.95791	valid-mae:1.95099
[1]	train-mae:1.67607	valid-mae:1.68316
[2]	train-mae:1.56875	valid-mae:1.58075
[3]	train-mae:1.50678	valid-mae:1.51842
[4]	train-mae:1.47096	valid-mae:1.48411
[5]	train-mae:1.44881	valid-mae:1.46061
[6]	train-mae:1.43352	valid-mae:1.44512
[7]	train-mae:1.41736	valid-mae:1.42981
[8]	train-mae:1.40822	valid-mae:1.42090
[9]	train-mae:1.39990	valid-mae:1.41455


[I 2022-05-14 19:43:52,637] Trial 22 finished with value: 1.414551593502439 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.25349966837529736, 'max_depth': 4, 'lambda': 0.8936043305784823, 'alpha': 0.006171548349129507, 'subsample': 0.9051007022054123, 'colsample_bytree': 0.9054820891046091, 'min_child_weight': 10, 'gamma': 0.0031516725846537654, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 1.0947580829056522e-08, 'skip_drop': 0.005023305228596638}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.29132	valid-mae:2.28137
[1]	train-mae:2.16626	valid-mae:2.15731
[2]	train-mae:2.07321	valid-mae:2.06571
[3]	train-mae:1.99017	valid-mae:1.98406
[4]	train-mae:1.91943	valid-mae:1.91615
[5]	train-mae:1.85946	valid-mae:1.85709
[6]	train-mae:1.80678	valid-mae:1.80565
[7]	train-mae:1.76052	valid-mae:1.76003
[8]	train-mae:1.71923	valid-mae:1.71892
[9]	train-mae:1.68125	valid-mae:1.68207


[I 2022-05-14 19:44:15,080] Trial 23 finished with value: 1.6820715149050174 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.0465237576029337, 'max_depth': 4, 'lambda': 0.17040269875004346, 'alpha': 4.918776039502122e-07, 'subsample': 0.9977604038883479, 'colsample_bytree': 0.9064323764364055, 'min_child_weight': 8, 'gamma': 0.11430883735347144, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 3.075778648025353e-06, 'skip_drop': 0.00010103561813388263}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.45533	valid-mae:2.45079
[1]	train-mae:2.38087	valid-mae:2.37587
[2]	train-mae:2.31972	valid-mae:2.31488
[3]	train-mae:2.26583	valid-mae:2.26151
[4]	train-mae:2.21933	valid-mae:2.21540
[5]	train-mae:2.17770	valid-mae:2.17435
[6]	train-mae:2.14047	valid-mae:2.13691
[7]	train-mae:2.10470	valid-mae:2.10199
[8]	train-mae:2.07182	valid-mae:2.06961
[9]	train-mae:2.04006	valid-mae:2.03771


[I 2022-05-14 19:44:30,339] Trial 24 finished with value: 2.0377122086766066 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.018211274221243676, 'max_depth': 4, 'lambda': 0.02107090671392615, 'alpha': 4.5418331780348525e-08, 'subsample': 0.8363970124988381, 'colsample_bytree': 0.7877254906317306, 'min_child_weight': 9, 'gamma': 0.03655873070440685, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 1.2429756612663286e-07, 'skip_drop': 0.000150632096292576}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.54232	valid-mae:2.53827
[1]	train-mae:2.53165	valid-mae:2.52743
[2]	train-mae:2.52125	valid-mae:2.51678
[3]	train-mae:2.51147	valid-mae:2.50686
[4]	train-mae:2.50227	valid-mae:2.49758
[5]	train-mae:2.49332	valid-mae:2.48854
[6]	train-mae:2.48441	valid-mae:2.47959
[7]	train-mae:2.47577	valid-mae:2.47098
[8]	train-mae:2.46757	valid-mae:2.46262
[9]	train-mae:2.45941	valid-mae:2.45434


[I 2022-05-14 19:44:47,106] Trial 25 finished with value: 2.4543408021399333 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.0017850806288132862, 'max_depth': 5, 'lambda': 0.0034656885147297334, 'alpha': 5.480938493901299e-06, 'subsample': 0.756161082829808, 'colsample_bytree': 0.6745012661825611, 'min_child_weight': 6, 'gamma': 0.9837274052987285, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 0.0002740395496100721, 'skip_drop': 1.0082597957831807e-08}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.12686	valid-mae:2.12805
[1]	train-mae:1.95181	valid-mae:1.95340
[2]	train-mae:1.82107	valid-mae:1.82413
[3]	train-mae:1.72344	valid-mae:1.72941
[4]	train-mae:1.65003	valid-mae:1.65779
[5]	train-mae:1.59295	valid-mae:1.60227
[6]	train-mae:1.54711	valid-mae:1.55649
[7]	train-mae:1.50906	valid-mae:1.51957
[8]	train-mae:1.47288	valid-mae:1.48535
[9]	train-mae:1.44653	valid-mae:1.46014


[I 2022-05-14 19:45:07,042] Trial 26 finished with value: 1.4601382346910246 and parameters: {'booster': 'gbtree', 'grow_policy': 'depthwise', 'eta': 0.09706181369516076, 'max_depth': 4, 'lambda': 0.23272292558214894, 'alpha': 5.474975216886147e-05, 'subsample': 0.8116013622705706, 'colsample_bytree': 0.8957639714920972, 'min_child_weight': 10, 'gamma': 0.0026664060178296697}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.54131	valid-mae:2.53724
[1]	train-mae:2.53080	valid-mae:2.52660
[2]	train-mae:2.52006	valid-mae:2.51575
[3]	train-mae:2.50948	valid-mae:2.50514
[4]	train-mae:2.49897	valid-mae:2.49450
[5]	train-mae:2.48828	valid-mae:2.48365
[6]	train-mae:2.47787	valid-mae:2.47312
[7]	train-mae:2.46795	valid-mae:2.46312
[8]	train-mae:2.45846	valid-mae:2.45349
[9]	train-mae:2.44911	valid-mae:2.44407


[I 2022-05-14 19:45:17,397] Trial 27 finished with value: 2.4440726504559622 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.002268537964996144, 'max_depth': 3, 'lambda': 0.00013394982610889263, 'alpha': 5.620516879158559e-07, 'subsample': 0.6633324606612153, 'colsample_bytree': 0.9531538532921585, 'min_child_weight': 8, 'gamma': 0.000416979613991516, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 2.7332790186617882e-05, 'skip_drop': 0.1936797559219304}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.13031	valid-mae:2.12105
[1]	train-mae:1.90169	valid-mae:1.89582
[2]	train-mae:1.76945	valid-mae:1.76832
[3]	train-mae:1.67832	valid-mae:1.67947
[4]	train-mae:1.60828	valid-mae:1.61155
[5]	train-mae:1.55373	valid-mae:1.55923
[6]	train-mae:1.51161	valid-mae:1.51832
[7]	train-mae:1.47888	valid-mae:1.48555
[8]	train-mae:1.45151	valid-mae:1.45883
[9]	train-mae:1.43001	valid-mae:1.43869


[I 2022-05-14 19:45:28,136] Trial 28 finished with value: 1.438688022977019 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.11430031126540571, 'max_depth': 4, 'lambda': 0.05975185133428871, 'alpha': 0.003994989302341713, 'subsample': 0.7264308512698215, 'colsample_bytree': 0.6168180519351647, 'min_child_weight': 7, 'gamma': 0.19850717541680218, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 3.492547448747176e-06, 'skip_drop': 0.009902807516474035}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.47123	valid-mae:2.46645
[1]	train-mae:2.39794	valid-mae:2.39292
[2]	train-mae:2.33915	valid-mae:2.33399
[3]	train-mae:2.28769	valid-mae:2.28238
[4]	train-mae:2.24337	valid-mae:2.23893
[5]	train-mae:2.20176	valid-mae:2.19789
[6]	train-mae:2.16457	valid-mae:2.16079
[7]	train-mae:2.12959	valid-mae:2.12642
[8]	train-mae:2.09830	valid-mae:2.09537
[9]	train-mae:2.06826	valid-mae:2.06565


[I 2022-05-14 19:45:54,232] Trial 29 finished with value: 2.065652137202828 and parameters: {'booster': 'dart', 'grow_policy': 'lossguide', 'eta': 0.0157557328701032, 'max_depth': 5, 'lambda': 0.000726237271203941, 'alpha': 3.913617261150293e-08, 'subsample': 0.8738561420778361, 'colsample_bytree': 0.8150156155648515, 'min_child_weight': 9, 'gamma': 0.016242393577260168, 'sample_type': 'weighted', 'normalize_type': 'tree', 'rate_drop': 1.0369403569981224e-07, 'skip_drop': 0.0026349526649547786}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.55009	valid-mae:2.54609
[1]	train-mae:2.54791	valid-mae:2.54386
[2]	train-mae:2.54571	valid-mae:2.54162
[3]	train-mae:2.54353	valid-mae:2.53941
[4]	train-mae:2.54139	valid-mae:2.53722
[5]	train-mae:2.53924	valid-mae:2.53504
[6]	train-mae:2.53712	valid-mae:2.53287
[7]	train-mae:2.53504	valid-mae:2.53076
[8]	train-mae:2.53299	valid-mae:2.52868
[9]	train-mae:2.53090	valid-mae:2.52657


[I 2022-05-14 19:46:01,232] Trial 30 finished with value: 2.5265733701351625 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.0003950426591439437, 'max_depth': 3, 'lambda': 0.9751067636664696, 'alpha': 2.6387442592934524e-06, 'subsample': 0.7846282816449024, 'colsample_bytree': 0.7200816813355513, 'min_child_weight': 5, 'gamma': 0.00158346800846877, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 0.0013986103637840604, 'skip_drop': 0.000429004843659837}. Best is trial 10 with value: 1.3931484339194795.


[0]	train-mae:2.06740	valid-mae:2.06664
[1]	train-mae:1.74873	valid-mae:1.74881
[2]	train-mae:1.62155	valid-mae:1.62492
[3]	train-mae:1.53876	valid-mae:1.54515
[4]	train-mae:1.49013	valid-mae:1.49785
[5]	train-mae:1.45193	valid-mae:1.46283
[6]	train-mae:1.42666	valid-mae:1.43596
[7]	train-mae:1.40748	valid-mae:1.41540
[8]	train-mae:1.39336	valid-mae:1.40220
[9]	train-mae:1.38148	valid-mae:1.39175


[I 2022-05-14 19:46:21,512] Trial 31 finished with value: 1.3917488294785456 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.19502384314948665, 'max_depth': 4, 'lambda': 0.3839413851896351, 'alpha': 3.271548114884581e-05, 'subsample': 0.8087637470753185, 'colsample_bytree': 0.9828590616563115, 'min_child_weight': 8, 'gamma': 0.010458740824139907, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 4.6579551666720177e-08, 'skip_drop': 0.00047103014517746846}. Best is trial 31 with value: 1.3917488294785456.


[0]	train-mae:2.17913	valid-mae:2.17568
[1]	train-mae:2.01836	valid-mae:2.01421
[2]	train-mae:1.89261	valid-mae:1.89191
[3]	train-mae:1.79718	valid-mae:1.79704
[4]	train-mae:1.71954	valid-mae:1.71985
[5]	train-mae:1.65733	valid-mae:1.65827
[6]	train-mae:1.60568	valid-mae:1.60816
[7]	train-mae:1.56188	valid-mae:1.56594
[8]	train-mae:1.52573	valid-mae:1.52966
[9]	train-mae:1.49542	valid-mae:1.50008


[I 2022-05-14 19:46:40,408] Trial 32 finished with value: 1.5000764658077335 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.08090710055762085, 'max_depth': 4, 'lambda': 0.26868108027544674, 'alpha': 0.0010320500846695067, 'subsample': 0.8560688954287942, 'colsample_bytree': 0.8784577944213809, 'min_child_weight': 8, 'gamma': 0.026381496544766913, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 2.095882204283117e-07, 'skip_drop': 0.00024228300245417313}. Best is trial 31 with value: 1.3917488294785456.


[0]	train-mae:2.08757	valid-mae:2.08887
[1]	train-mae:1.65298	valid-mae:1.65883
[2]	train-mae:1.56863	valid-mae:1.57947
[3]	train-mae:1.51660	valid-mae:1.53234
[4]	train-mae:1.48589	valid-mae:1.50123
[5]	train-mae:1.46804	valid-mae:1.48221
[6]	train-mae:1.45359	valid-mae:1.46571
[7]	train-mae:1.44081	valid-mae:1.45633
[8]	train-mae:1.42942	valid-mae:1.44336
[9]	train-mae:1.42062	valid-mae:1.43497


[I 2022-05-14 19:47:00,169] Trial 33 finished with value: 1.4349652640345414 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.28860248215611395, 'max_depth': 4, 'lambda': 0.1118660049771641, 'alpha': 8.437149745772984e-07, 'subsample': 0.8001147323735756, 'colsample_bytree': 0.9554251321737344, 'min_child_weight': 9, 'gamma': 0.004823799005573686, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 6.736875734378963e-08, 'skip_drop': 0.04954013388336943}. Best is trial 31 with value: 1.3917488294785456.


[0]	train-mae:2.47275	valid-mae:2.46754
[1]	train-mae:2.40669	valid-mae:2.40076
[2]	train-mae:2.35203	valid-mae:2.34578
[3]	train-mae:2.30363	valid-mae:2.29718
[4]	train-mae:2.26042	valid-mae:2.25499
[5]	train-mae:2.22175	valid-mae:2.21634
[6]	train-mae:2.18736	valid-mae:2.18224
[7]	train-mae:2.15568	valid-mae:2.15078
[8]	train-mae:2.12602	valid-mae:2.12155
[9]	train-mae:2.09640	valid-mae:2.09281


[I 2022-05-14 19:47:21,060] Trial 34 finished with value: 2.092813058478462 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.01499979537653951, 'max_depth': 4, 'lambda': 0.013818483324499156, 'alpha': 8.326664780812847e-05, 'subsample': 0.9040319162944533, 'colsample_bytree': 0.9316015170353585, 'min_child_weight': 8, 'gamma': 9.915138177676405e-05, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 1.4751516145946865e-06, 'skip_drop': 1.8693160198951176e-05}. Best is trial 31 with value: 1.3917488294785456.


[0]	train-mae:2.55233	valid-mae:2.54836
[1]	train-mae:2.55233	valid-mae:2.54836
[2]	train-mae:2.55233	valid-mae:2.54836
[3]	train-mae:2.55233	valid-mae:2.54836
[4]	train-mae:2.55233	valid-mae:2.54836
[5]	train-mae:2.55233	valid-mae:2.54836
[6]	train-mae:2.55233	valid-mae:2.54836
[7]	train-mae:2.55233	valid-mae:2.54836
[8]	train-mae:2.55233	valid-mae:2.54836
[9]	train-mae:2.55233	valid-mae:2.54836


[I 2022-05-14 19:47:49,856] Trial 35 finished with value: 2.54835913505779 and parameters: {'booster': 'gbtree', 'grow_policy': 'depthwise', 'eta': 8.597139790329569e-08, 'max_depth': 5, 'lambda': 0.46516881147299, 'alpha': 4.963594829092325e-06, 'subsample': 0.7363220363813129, 'colsample_bytree': 0.9991813199418469, 'min_child_weight': 10, 'gamma': 0.08229518037244679}. Best is trial 31 with value: 1.3917488294785456.


[0]	train-mae:2.10936	valid-mae:2.09784
[1]	train-mae:1.91101	valid-mae:1.90425
[2]	train-mae:1.77966	valid-mae:1.77505
[3]	train-mae:1.67943	valid-mae:1.67977
[4]	train-mae:1.60294	valid-mae:1.60866
[5]	train-mae:1.54436	valid-mae:1.55225
[6]	train-mae:1.49796	valid-mae:1.50738
[7]	train-mae:1.46106	valid-mae:1.47344
[8]	train-mae:1.42932	valid-mae:1.44174
[9]	train-mae:1.40368	valid-mae:1.41663


[I 2022-05-14 19:48:31,464] Trial 36 finished with value: 1.416631138751249 and parameters: {'booster': 'dart', 'grow_policy': 'lossguide', 'eta': 0.10203070738922489, 'max_depth': 6, 'lambda': 0.019212186950187123, 'alpha': 0.030693950692783467, 'subsample': 0.8381614082421027, 'colsample_bytree': 0.868488455753744, 'min_child_weight': 7, 'gamma': 0.00023342075594307596, 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 4.3499430966000124e-08, 'skip_drop': 0.0045561401002233}. Best is trial 31 with value: 1.3917488294785456.


[0]	train-mae:2.53498	valid-mae:2.53099
[1]	train-mae:2.51721	valid-mae:2.51322
[2]	train-mae:2.49881	valid-mae:2.49466
[3]	train-mae:2.48047	valid-mae:2.47616
[4]	train-mae:2.46333	valid-mae:2.45892
[5]	train-mae:2.44728	valid-mae:2.44272
[6]	train-mae:2.43229	valid-mae:2.42769
[7]	train-mae:2.41784	valid-mae:2.41318
[8]	train-mae:2.40396	valid-mae:2.39920
[9]	train-mae:2.39069	valid-mae:2.38585


[I 2022-05-14 19:48:39,862] Trial 37 finished with value: 2.3858484776793545 and parameters: {'booster': 'gbtree', 'grow_policy': 'depthwise', 'eta': 0.0036982863724254293, 'max_depth': 4, 'lambda': 4.177606091377989e-06, 'alpha': 2.3549282239963276e-07, 'subsample': 0.6939722268719406, 'colsample_bytree': 0.5431450072158877, 'min_child_weight': 8, 'gamma': 0.32069999817484446}. Best is trial 31 with value: 1.3917488294785456.


[0]	train-mae:2.39551	valid-mae:2.38929
[1]	train-mae:2.29520	valid-mae:2.28954
[2]	train-mae:2.21119	valid-mae:2.20675
[3]	train-mae:2.14336	valid-mae:2.14063
[4]	train-mae:2.08682	valid-mae:2.08444
[5]	train-mae:2.03288	valid-mae:2.03165
[6]	train-mae:1.98414	valid-mae:1.98372
[7]	train-mae:1.93844	valid-mae:1.93920
[8]	train-mae:1.89501	valid-mae:1.89615
[9]	train-mae:1.85669	valid-mae:1.85811


[I 2022-05-14 19:49:10,162] Trial 38 finished with value: 1.85811438459204 and parameters: {'booster': 'dart', 'grow_policy': 'depthwise', 'eta': 0.0280117054049344, 'max_depth': 5, 'lambda': 0.04046810054265252, 'alpha': 4.1319871454466696e-05, 'subsample': 0.7734850314676558, 'colsample_bytree': 0.8038129584607341, 'min_child_weight': 6, 'gamma': 3.6662998865747404e-06, 'sample_type': 'uniform', 'normalize_type': 'tree', 'rate_drop': 1.3206848783228852e-05, 'skip_drop': 5.130919231681643e-06}. Best is trial 31 with value: 1.3917488294785456.


Number of finished trials:  39
Best trial:
  Value: 1.3917488294785456
  Params: 
    booster: dart
    grow_policy: depthwise
    eta: 0.19502384314948665
    max_depth: 4
    lambda: 0.3839413851896351
    alpha: 3.271548114884581e-05
    subsample: 0.8087637470753185
    colsample_bytree: 0.9828590616563115
    min_child_weight: 8
    gamma: 0.010458740824139907
    sample_type: uniform
    normalize_type: forest
    rate_drop: 4.6579551666720177e-08
    skip_drop: 0.00047103014517746846


Value: 1.3917488294785456  
  Params:   
    booster: dart  
    grow_policy: depthwise  
    eta: 0.19502384314948665  
    max_depth: 4  
    lambda: 0.3839413851896351  
    alpha: 3.271548114884581e-05  
    subsample: 0.8087637470753185  
    colsample_bytree: 0.9828590616563115  
    min_child_weight: 8  
    gamma: 0.010458740824139907  
    sample_type: uniform  
    normalize_type: forest  
    rate_drop: 4.6579551666720177e-08  
    skip_drop: 0.00047103014517746846  

### Tuning # 2

In [9]:
def optimize(trial, X, y):
    train_x, valid_x, train_y, valid_y = train_test_split(X, y, test_size=0.20, random_state=SEED)

    dtrain = xgb.DMatrix(train_x, label=train_y)
    dvalid = xgb.DMatrix(valid_x, label=valid_y)

    param = {
        "seed": SEED,
        "verbosity": 0,
        "objective": "reg:pseudohubererror",
        "eval_metric": "mae",
        # gpu_hist for faster training
        "tree_method": "gpu_hist",
        # defines booster, gblinear for linear functions.
        "booster": "dart",
        # defines how selective algorithm is.
        "grow_policy": "depthwise",
        # learning rate should be tuned also
        "eta": trial.suggest_float("eta", 0.01, 0.3, log=True),
        # the maximum depth of a tree
        "max_depth": 4,
        # L2 regularization weight.
        "lambda": 0.3839413851896351,
        # L1 regularization weight.
        "alpha": 3.271548114884581e-05,
        # sampling ratio for training data.
        "subsample": 0.8087637470753185,
        # sampling according to each tree.
        "colsample_bytree": 0.9828590616563115,
        # sampling for each split.
        "colsample_bylevel": 0.9828590616563115,
        # minimum child weight, larger the term more conservative the tree.
        "min_child_weight": 8,
        # defines how selective algorithm is.
        "gamma": 0.010458740824139907,
    }

    param["sample_type"] = "uniform"
    param["normalize_type"] = "forest"
    param["rate_drop"] = 4.6579551666720177e-08
    param["skip_drop"] = 0.00047103014517746846

    bst = xgb.train(param, dtrain, num_boost_round=100, evals=[(dtrain, 'train') , (dvalid, 'valid')])
    preds = bst.predict(dvalid)
    mae = mean_absolute_error(valid_y, preds)
    return mae

In [11]:
study = optuna.create_study(direction="minimize")
optimization_function = partial(optimize, X=X_train_sparse, y=y_train)
study.optimize(optimization_function, n_trials=100, timeout=6000)

print("Number of finished trials: ", len(study.trials))
print("Best trial:")

trial = study.best_trial

print("  Value: {}".format(trial.value))
print("  Params: ")

for key, value in trial.params.items():
    print("    {}: {}".format(key, value))

[I 2022-05-14 20:15:53,117] A new study created in memory with name: no-name-994d91fb-9fe2-4d66-97bd-94a4af889299


[0]	train-mae:2.25237	valid-mae:2.24447
[1]	train-mae:2.11534	valid-mae:2.10749
[2]	train-mae:2.00908	valid-mae:2.00286
[3]	train-mae:1.91505	valid-mae:1.90999
[4]	train-mae:1.84119	valid-mae:1.83661
[5]	train-mae:1.77773	valid-mae:1.77424
[6]	train-mae:1.72232	valid-mae:1.71854
[7]	train-mae:1.67524	valid-mae:1.67262
[8]	train-mae:1.63467	valid-mae:1.63367
[9]	train-mae:1.59869	valid-mae:1.59921
[10]	train-mae:1.56618	valid-mae:1.56766
[11]	train-mae:1.53846	valid-mae:1.54136
[12]	train-mae:1.51342	valid-mae:1.51723
[13]	train-mae:1.49189	valid-mae:1.49569
[14]	train-mae:1.47275	valid-mae:1.47695
[15]	train-mae:1.45591	valid-mae:1.46106
[16]	train-mae:1.44002	valid-mae:1.44582
[17]	train-mae:1.42630	valid-mae:1.43286
[18]	train-mae:1.41397	valid-mae:1.42059
[19]	train-mae:1.40290	valid-mae:1.40993
[20]	train-mae:1.39297	valid-mae:1.40060
[21]	train-mae:1.38396	valid-mae:1.39207
[22]	train-mae:1.37523	valid-mae:1.38319
[23]	train-mae:1.36830	valid-mae:1.37642
[24]	train-mae:1.36158	val

[I 2022-05-14 20:19:56,110] Trial 0 finished with value: 1.2688200989606697 and parameters: {'eta': 0.05853850936494535}. Best is trial 0 with value: 1.2688200989606697.


[0]	train-mae:2.13170	valid-mae:2.11972
[1]	train-mae:1.94450	valid-mae:1.93430
[2]	train-mae:1.81546	valid-mae:1.80820
[3]	train-mae:1.71823	valid-mae:1.71460
[4]	train-mae:1.64527	valid-mae:1.64339
[5]	train-mae:1.58804	valid-mae:1.58746
[6]	train-mae:1.54279	valid-mae:1.54326
[7]	train-mae:1.50722	valid-mae:1.50903
[8]	train-mae:1.47782	valid-mae:1.48207
[9]	train-mae:1.45407	valid-mae:1.45831
[10]	train-mae:1.43341	valid-mae:1.43835
[11]	train-mae:1.41597	valid-mae:1.42186
[12]	train-mae:1.40207	valid-mae:1.40786
[13]	train-mae:1.38876	valid-mae:1.39514
[14]	train-mae:1.37769	valid-mae:1.38518
[15]	train-mae:1.36973	valid-mae:1.37811
[16]	train-mae:1.36241	valid-mae:1.37107
[17]	train-mae:1.35656	valid-mae:1.36543
[18]	train-mae:1.35137	valid-mae:1.36126
[19]	train-mae:1.34538	valid-mae:1.35609
[20]	train-mae:1.34055	valid-mae:1.35231
[21]	train-mae:1.33603	valid-mae:1.34814
[22]	train-mae:1.33230	valid-mae:1.34489
[23]	train-mae:1.32859	valid-mae:1.34233
[24]	train-mae:1.32467	val

[I 2022-05-14 20:22:13,620] Trial 1 finished with value: 1.2436840722598155 and parameters: {'eta': 0.10353223325237743}. Best is trial 1 with value: 1.2436840722598155.


[0]	train-mae:2.49046	valid-mae:2.48565
[1]	train-mae:2.43669	valid-mae:2.43139
[2]	train-mae:2.39169	valid-mae:2.38582
[3]	train-mae:2.35008	valid-mae:2.34395
[4]	train-mae:2.31277	valid-mae:2.30682
[5]	train-mae:2.27690	valid-mae:2.27143
[6]	train-mae:2.24597	valid-mae:2.24050
[7]	train-mae:2.21630	valid-mae:2.21117
[8]	train-mae:2.18930	valid-mae:2.18438
[9]	train-mae:2.16438	valid-mae:2.15953
[10]	train-mae:2.14109	valid-mae:2.13643
[11]	train-mae:2.11784	valid-mae:2.11344
[12]	train-mae:2.09607	valid-mae:2.09219
[13]	train-mae:2.07475	valid-mae:2.07090
[14]	train-mae:2.05443	valid-mae:2.05086
[15]	train-mae:2.03442	valid-mae:2.03124
[16]	train-mae:2.01483	valid-mae:2.01193
[17]	train-mae:1.99580	valid-mae:1.99317
[18]	train-mae:1.97721	valid-mae:1.97453
[19]	train-mae:1.95927	valid-mae:1.95662
[20]	train-mae:1.94168	valid-mae:1.93916
[21]	train-mae:1.92516	valid-mae:1.92298
[22]	train-mae:1.90964	valid-mae:1.90766
[23]	train-mae:1.89441	valid-mae:1.89241
[24]	train-mae:1.87942	val

[I 2022-05-14 20:24:32,372] Trial 2 finished with value: 1.4160889979356779 and parameters: {'eta': 0.011649868248131797}. Best is trial 1 with value: 1.2436840722598155.


[0]	train-mae:2.25031	valid-mae:2.24233
[1]	train-mae:2.11274	valid-mae:2.10481
[2]	train-mae:2.00612	valid-mae:1.99982
[3]	train-mae:1.91202	valid-mae:1.90687
[4]	train-mae:1.83809	valid-mae:1.83342
[5]	train-mae:1.77461	valid-mae:1.77102
[6]	train-mae:1.71924	valid-mae:1.71536
[7]	train-mae:1.67237	valid-mae:1.66969
[8]	train-mae:1.63189	valid-mae:1.63084
[9]	train-mae:1.59603	valid-mae:1.59649
[10]	train-mae:1.56362	valid-mae:1.56461
[11]	train-mae:1.53571	valid-mae:1.53799
[12]	train-mae:1.51130	valid-mae:1.51408
[13]	train-mae:1.48944	valid-mae:1.49230
[14]	train-mae:1.47004	valid-mae:1.47398
[15]	train-mae:1.45337	valid-mae:1.45831
[16]	train-mae:1.43818	valid-mae:1.44378
[17]	train-mae:1.42499	valid-mae:1.43125
[18]	train-mae:1.41276	valid-mae:1.41941
[19]	train-mae:1.40144	valid-mae:1.40879
[20]	train-mae:1.39092	valid-mae:1.39854
[21]	train-mae:1.38256	valid-mae:1.39021
[22]	train-mae:1.37442	valid-mae:1.38286
[23]	train-mae:1.36730	valid-mae:1.37652
[24]	train-mae:1.36054	val

[I 2022-05-14 20:26:52,440] Trial 3 finished with value: 1.2671695635702835 and parameters: {'eta': 0.059034259243055165}. Best is trial 1 with value: 1.2436840722598155.


[0]	train-mae:2.43694	valid-mae:2.43130
[1]	train-mae:2.35437	valid-mae:2.34801
[2]	train-mae:2.28433	valid-mae:2.27766
[3]	train-mae:2.22552	valid-mae:2.21947
[4]	train-mae:2.17538	valid-mae:2.16926
[5]	train-mae:2.12913	valid-mae:2.12404
[6]	train-mae:2.08781	valid-mae:2.08321
[7]	train-mae:2.04894	valid-mae:2.04484
[8]	train-mae:2.01230	valid-mae:2.00861
[9]	train-mae:1.97680	valid-mae:1.97333
[10]	train-mae:1.94383	valid-mae:1.94044
[11]	train-mae:1.91293	valid-mae:1.90988
[12]	train-mae:1.88448	valid-mae:1.88178
[13]	train-mae:1.85754	valid-mae:1.85529
[14]	train-mae:1.83180	valid-mae:1.83043
[15]	train-mae:1.80785	valid-mae:1.80700
[16]	train-mae:1.78494	valid-mae:1.78463
[17]	train-mae:1.76304	valid-mae:1.76338
[18]	train-mae:1.74252	valid-mae:1.74354
[19]	train-mae:1.72321	valid-mae:1.72480
[20]	train-mae:1.70471	valid-mae:1.70714
[21]	train-mae:1.68723	valid-mae:1.69026
[22]	train-mae:1.67115	valid-mae:1.67491
[23]	train-mae:1.65567	valid-mae:1.65994
[24]	train-mae:1.64094	val

[I 2022-05-14 20:29:14,546] Trial 4 finished with value: 1.3265298130142955 and parameters: {'eta': 0.021777116944679757}. Best is trial 1 with value: 1.2436840722598155.


[0]	train-mae:2.46895	valid-mae:2.46384
[1]	train-mae:2.40227	valid-mae:2.39652
[2]	train-mae:2.34600	valid-mae:2.33993
[3]	train-mae:2.29598	valid-mae:2.29023
[4]	train-mae:2.25201	valid-mae:2.24647
[5]	train-mae:2.21134	valid-mae:2.20625
[6]	train-mae:2.17536	valid-mae:2.17068
[7]	train-mae:2.14318	valid-mae:2.13923
[8]	train-mae:2.11240	valid-mae:2.10887
[9]	train-mae:2.08196	valid-mae:2.07866
[10]	train-mae:2.05405	valid-mae:2.05091
[11]	train-mae:2.02666	valid-mae:2.02414
[12]	train-mae:2.00139	valid-mae:1.99919
[13]	train-mae:1.97681	valid-mae:1.97487
[14]	train-mae:1.95265	valid-mae:1.95137
[15]	train-mae:1.93026	valid-mae:1.92952
[16]	train-mae:1.90877	valid-mae:1.90821
[17]	train-mae:1.88836	valid-mae:1.88799
[18]	train-mae:1.86883	valid-mae:1.86865
[19]	train-mae:1.85013	valid-mae:1.85049
[20]	train-mae:1.83187	valid-mae:1.83241
[21]	train-mae:1.81414	valid-mae:1.81502
[22]	train-mae:1.79727	valid-mae:1.79850
[23]	train-mae:1.78124	valid-mae:1.78291
[24]	train-mae:1.76546	val

[I 2022-05-14 20:31:32,995] Trial 5 finished with value: 1.3629676819780747 and parameters: {'eta': 0.0157149991224597}. Best is trial 1 with value: 1.2436840722598155.


[0]	train-mae:2.49283	valid-mae:2.48805
[1]	train-mae:2.44058	valid-mae:2.43532
[2]	train-mae:2.39680	valid-mae:2.39098
[3]	train-mae:2.35644	valid-mae:2.35026
[4]	train-mae:2.31964	valid-mae:2.31375
[5]	train-mae:2.28383	valid-mae:2.27829
[6]	train-mae:2.25289	valid-mae:2.24762
[7]	train-mae:2.22402	valid-mae:2.21918
[8]	train-mae:2.19780	valid-mae:2.19313
[9]	train-mae:2.17343	valid-mae:2.16894
[10]	train-mae:2.15063	valid-mae:2.14640
[11]	train-mae:2.12777	valid-mae:2.12367
[12]	train-mae:2.10604	valid-mae:2.10208
[13]	train-mae:2.08495	valid-mae:2.08116
[14]	train-mae:2.06505	valid-mae:2.06153
[15]	train-mae:2.04568	valid-mae:2.04256
[16]	train-mae:2.02625	valid-mae:2.02330
[17]	train-mae:2.00754	valid-mae:2.00479
[18]	train-mae:1.98944	valid-mae:1.98678
[19]	train-mae:1.97221	valid-mae:1.96960
[20]	train-mae:1.95485	valid-mae:1.95247
[21]	train-mae:1.93843	valid-mae:1.93635
[22]	train-mae:1.92274	valid-mae:1.92075
[23]	train-mae:1.90746	valid-mae:1.90570
[24]	train-mae:1.89275	val

[I 2022-05-14 20:33:50,887] Trial 6 finished with value: 1.4252473280911926 and parameters: {'eta': 0.011203643501683298}. Best is trial 1 with value: 1.2436840722598155.


[0]	train-mae:2.19667	valid-mae:2.18648
[1]	train-mae:2.04897	valid-mae:2.03935
[2]	train-mae:1.93033	valid-mae:1.92262
[3]	train-mae:1.83747	valid-mae:1.83088
[4]	train-mae:1.76223	valid-mae:1.75477
[5]	train-mae:1.69929	valid-mae:1.69227
[6]	train-mae:1.64776	valid-mae:1.64166
[7]	train-mae:1.60391	valid-mae:1.60052
[8]	train-mae:1.56620	valid-mae:1.56481
[9]	train-mae:1.53372	valid-mae:1.53360
[10]	train-mae:1.50376	valid-mae:1.50454
[11]	train-mae:1.48057	valid-mae:1.48274
[12]	train-mae:1.46086	valid-mae:1.46362
[13]	train-mae:1.44363	valid-mae:1.44696
[14]	train-mae:1.42783	valid-mae:1.43232
[15]	train-mae:1.41346	valid-mae:1.41854
[16]	train-mae:1.40127	valid-mae:1.40775
[17]	train-mae:1.39026	valid-mae:1.39750
[18]	train-mae:1.38087	valid-mae:1.38877
[19]	train-mae:1.37309	valid-mae:1.38085
[20]	train-mae:1.36572	valid-mae:1.37385
[21]	train-mae:1.35941	valid-mae:1.36748
[22]	train-mae:1.35346	valid-mae:1.36212
[23]	train-mae:1.34765	valid-mae:1.35702
[24]	train-mae:1.34227	val

[I 2022-05-14 20:36:08,226] Trial 7 finished with value: 1.2602819460008456 and parameters: {'eta': 0.07327237393111315}. Best is trial 1 with value: 1.2436840722598155.


[0]	train-mae:2.06589	valid-mae:2.06906
[1]	train-mae:1.69047	valid-mae:1.68827
[2]	train-mae:1.57147	valid-mae:1.56698
[3]	train-mae:1.50804	valid-mae:1.50893
[4]	train-mae:1.46690	valid-mae:1.47264
[5]	train-mae:1.44166	valid-mae:1.44890
[6]	train-mae:1.42462	valid-mae:1.43238
[7]	train-mae:1.41014	valid-mae:1.41651
[8]	train-mae:1.40086	valid-mae:1.40953
[9]	train-mae:1.39192	valid-mae:1.40114
[10]	train-mae:1.38511	valid-mae:1.39515
[11]	train-mae:1.37619	valid-mae:1.38681
[12]	train-mae:1.36749	valid-mae:1.37860
[13]	train-mae:1.36011	valid-mae:1.37336
[14]	train-mae:1.35504	valid-mae:1.36954
[15]	train-mae:1.35022	valid-mae:1.36622
[16]	train-mae:1.34469	valid-mae:1.36180
[17]	train-mae:1.33846	valid-mae:1.35624
[18]	train-mae:1.33364	valid-mae:1.35285
[19]	train-mae:1.32791	valid-mae:1.34844
[20]	train-mae:1.32360	valid-mae:1.34610
[21]	train-mae:1.31988	valid-mae:1.34255
[22]	train-mae:1.31604	valid-mae:1.34009
[23]	train-mae:1.31264	valid-mae:1.33812
[24]	train-mae:1.30919	val

[I 2022-05-14 20:38:19,138] Trial 8 finished with value: 1.2473300555948845 and parameters: {'eta': 0.23534261543615248}. Best is trial 1 with value: 1.2436840722598155.


[0]	train-mae:2.37155	valid-mae:2.36490
[1]	train-mae:2.26109	valid-mae:2.25526
[2]	train-mae:2.17804	valid-mae:2.17239
[3]	train-mae:2.10690	valid-mae:2.10177
[4]	train-mae:2.04372	valid-mae:2.03942
[5]	train-mae:1.98597	valid-mae:1.98283
[6]	train-mae:1.93397	valid-mae:1.93221
[7]	train-mae:1.88778	valid-mae:1.88652
[8]	train-mae:1.84523	valid-mae:1.84473
[9]	train-mae:1.80638	valid-mae:1.80645
[10]	train-mae:1.77118	valid-mae:1.77219
[11]	train-mae:1.73829	valid-mae:1.74021
[12]	train-mae:1.70818	valid-mae:1.71103
[13]	train-mae:1.68110	valid-mae:1.68465
[14]	train-mae:1.65593	valid-mae:1.66044
[15]	train-mae:1.63284	valid-mae:1.63799
[16]	train-mae:1.61161	valid-mae:1.61789
[17]	train-mae:1.59210	valid-mae:1.59893
[18]	train-mae:1.57127	valid-mae:1.57905
[19]	train-mae:1.55408	valid-mae:1.56225
[20]	train-mae:1.53827	valid-mae:1.54671
[21]	train-mae:1.52335	valid-mae:1.53187
[22]	train-mae:1.50988	valid-mae:1.51831
[23]	train-mae:1.49694	valid-mae:1.50557
[24]	train-mae:1.48465	val

[I 2022-05-14 20:40:40,846] Trial 9 finished with value: 1.295941191507906 and parameters: {'eta': 0.03421058118353234}. Best is trial 1 with value: 1.2436840722598155.


[0]	train-mae:2.08396	valid-mae:2.07837
[1]	train-mae:1.81355	valid-mae:1.81216
[2]	train-mae:1.68206	valid-mae:1.68453
[3]	train-mae:1.59297	valid-mae:1.59862
[4]	train-mae:1.52783	valid-mae:1.53646
[5]	train-mae:1.48219	valid-mae:1.48967
[6]	train-mae:1.44933	valid-mae:1.45502
[7]	train-mae:1.42615	valid-mae:1.43141
[8]	train-mae:1.40655	valid-mae:1.41039
[9]	train-mae:1.38935	valid-mae:1.39420
[10]	train-mae:1.37878	valid-mae:1.38508
[11]	train-mae:1.36954	valid-mae:1.37688
[12]	train-mae:1.36157	valid-mae:1.37048
[13]	train-mae:1.35445	valid-mae:1.36385
[14]	train-mae:1.34851	valid-mae:1.35885
[15]	train-mae:1.34233	valid-mae:1.35396
[16]	train-mae:1.33747	valid-mae:1.34933
[17]	train-mae:1.33308	valid-mae:1.34646
[18]	train-mae:1.32800	valid-mae:1.34233
[19]	train-mae:1.32329	valid-mae:1.33747
[20]	train-mae:1.31968	valid-mae:1.33465
[21]	train-mae:1.31691	valid-mae:1.33226
[22]	train-mae:1.31289	valid-mae:1.32884
[23]	train-mae:1.30926	valid-mae:1.32588
[24]	train-mae:1.30592	val

[I 2022-05-14 20:42:54,263] Trial 10 finished with value: 1.234455828166958 and parameters: {'eta': 0.15406470353245602}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.08152	valid-mae:2.07644
[1]	train-mae:1.80776	valid-mae:1.80660
[2]	train-mae:1.67801	valid-mae:1.67983
[3]	train-mae:1.58858	valid-mae:1.59500
[4]	train-mae:1.52478	valid-mae:1.53321
[5]	train-mae:1.47899	valid-mae:1.48839
[6]	train-mae:1.44651	valid-mae:1.45456
[7]	train-mae:1.42337	valid-mae:1.43124
[8]	train-mae:1.40459	valid-mae:1.41241
[9]	train-mae:1.39162	valid-mae:1.40012
[10]	train-mae:1.37884	valid-mae:1.38851
[11]	train-mae:1.36967	valid-mae:1.38035
[12]	train-mae:1.36202	valid-mae:1.37352
[13]	train-mae:1.35542	valid-mae:1.36815
[14]	train-mae:1.35075	valid-mae:1.36357
[15]	train-mae:1.34434	valid-mae:1.35741
[16]	train-mae:1.33955	valid-mae:1.35268
[17]	train-mae:1.33458	valid-mae:1.34894
[18]	train-mae:1.33041	valid-mae:1.34495
[19]	train-mae:1.32671	valid-mae:1.34256
[20]	train-mae:1.32300	valid-mae:1.33933
[21]	train-mae:1.32060	valid-mae:1.33757
[22]	train-mae:1.31714	valid-mae:1.33486
[23]	train-mae:1.31272	valid-mae:1.33095
[24]	train-mae:1.30861	val

[I 2022-05-14 20:45:05,812] Trial 11 finished with value: 1.2411927547136707 and parameters: {'eta': 0.1576351910360057}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.06562	valid-mae:2.06651
[1]	train-mae:1.72392	valid-mae:1.72280
[2]	train-mae:1.58818	valid-mae:1.58645
[3]	train-mae:1.51376	valid-mae:1.51370
[4]	train-mae:1.46497	valid-mae:1.46956
[5]	train-mae:1.43447	valid-mae:1.43978
[6]	train-mae:1.41421	valid-mae:1.41973
[7]	train-mae:1.39877	valid-mae:1.40469
[8]	train-mae:1.38771	valid-mae:1.39526
[9]	train-mae:1.37696	valid-mae:1.38472
[10]	train-mae:1.36745	valid-mae:1.37676
[11]	train-mae:1.35757	valid-mae:1.36731
[12]	train-mae:1.35141	valid-mae:1.36193
[13]	train-mae:1.34571	valid-mae:1.35779
[14]	train-mae:1.34048	valid-mae:1.35418
[15]	train-mae:1.33535	valid-mae:1.35107
[16]	train-mae:1.32989	valid-mae:1.34596
[17]	train-mae:1.32436	valid-mae:1.34156
[18]	train-mae:1.31994	valid-mae:1.33847
[19]	train-mae:1.31557	valid-mae:1.33488
[20]	train-mae:1.31199	valid-mae:1.33213
[21]	train-mae:1.30915	valid-mae:1.32967
[22]	train-mae:1.30592	valid-mae:1.32625
[23]	train-mae:1.30179	valid-mae:1.32322
[24]	train-mae:1.29774	val

[I 2022-05-14 20:47:15,284] Trial 12 finished with value: 1.2388688056519084 and parameters: {'eta': 0.21112444181178402}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.07222	valid-mae:2.07746
[1]	train-mae:1.66098	valid-mae:1.65550
[2]	train-mae:1.55419	valid-mae:1.54779
[3]	train-mae:1.50053	valid-mae:1.50096
[4]	train-mae:1.46604	valid-mae:1.46569
[5]	train-mae:1.44268	valid-mae:1.44172
[6]	train-mae:1.42396	valid-mae:1.42452
[7]	train-mae:1.41148	valid-mae:1.41418
[8]	train-mae:1.40110	valid-mae:1.40576
[9]	train-mae:1.39035	valid-mae:1.39777
[10]	train-mae:1.38200	valid-mae:1.39090
[11]	train-mae:1.37550	valid-mae:1.38542
[12]	train-mae:1.36937	valid-mae:1.38160
[13]	train-mae:1.36295	valid-mae:1.37667
[14]	train-mae:1.35765	valid-mae:1.37178
[15]	train-mae:1.35029	valid-mae:1.36606
[16]	train-mae:1.34199	valid-mae:1.35922
[17]	train-mae:1.33570	valid-mae:1.35427
[18]	train-mae:1.33044	valid-mae:1.34840
[19]	train-mae:1.32666	valid-mae:1.34612
[20]	train-mae:1.32111	valid-mae:1.34194
[21]	train-mae:1.31712	valid-mae:1.33875
[22]	train-mae:1.31230	valid-mae:1.33478
[23]	train-mae:1.30787	valid-mae:1.33261
[24]	train-mae:1.30359	val

[I 2022-05-14 20:49:24,123] Trial 13 finished with value: 1.2438723762227601 and parameters: {'eta': 0.2697409453234533}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.10087	valid-mae:2.09229
[1]	train-mae:1.89376	valid-mae:1.89452
[2]	train-mae:1.72348	valid-mae:1.72072
[3]	train-mae:1.63229	valid-mae:1.63402
[4]	train-mae:1.56588	valid-mae:1.57113
[5]	train-mae:1.51544	valid-mae:1.52409
[6]	train-mae:1.47909	valid-mae:1.48793
[7]	train-mae:1.45072	valid-mae:1.45928
[8]	train-mae:1.42817	valid-mae:1.43712
[9]	train-mae:1.40978	valid-mae:1.41881
[10]	train-mae:1.39515	valid-mae:1.40506
[11]	train-mae:1.38243	valid-mae:1.39271
[12]	train-mae:1.37347	valid-mae:1.38398
[13]	train-mae:1.36468	valid-mae:1.37676
[14]	train-mae:1.35654	valid-mae:1.37014
[15]	train-mae:1.34823	valid-mae:1.36167
[16]	train-mae:1.34164	valid-mae:1.35523
[17]	train-mae:1.33712	valid-mae:1.35272
[18]	train-mae:1.33333	valid-mae:1.34914
[19]	train-mae:1.32897	valid-mae:1.34525
[20]	train-mae:1.32437	valid-mae:1.34208
[21]	train-mae:1.32090	valid-mae:1.33927
[22]	train-mae:1.31787	valid-mae:1.33667
[23]	train-mae:1.31512	valid-mae:1.33419
[24]	train-mae:1.31101	val

[I 2022-05-14 20:51:39,475] Trial 14 finished with value: 1.2345525404578657 and parameters: {'eta': 0.13262768166900787}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.11460	valid-mae:2.10388
[1]	train-mae:1.88773	valid-mae:1.88255
[2]	train-mae:1.75961	valid-mae:1.75789
[3]	train-mae:1.66839	valid-mae:1.67054
[4]	train-mae:1.59991	valid-mae:1.60482
[5]	train-mae:1.54564	valid-mae:1.55050
[6]	train-mae:1.50579	valid-mae:1.51102
[7]	train-mae:1.47406	valid-mae:1.48063
[8]	train-mae:1.44750	valid-mae:1.45548
[9]	train-mae:1.42623	valid-mae:1.43483
[10]	train-mae:1.40842	valid-mae:1.41913
[11]	train-mae:1.39293	valid-mae:1.40375
[12]	train-mae:1.37820	valid-mae:1.39038
[13]	train-mae:1.36814	valid-mae:1.38140
[14]	train-mae:1.35942	valid-mae:1.37448
[15]	train-mae:1.35231	valid-mae:1.36785
[16]	train-mae:1.34572	valid-mae:1.36181
[17]	train-mae:1.33998	valid-mae:1.35690
[18]	train-mae:1.33565	valid-mae:1.35258
[19]	train-mae:1.33145	valid-mae:1.34878
[20]	train-mae:1.32730	valid-mae:1.34498
[21]	train-mae:1.32295	valid-mae:1.34149
[22]	train-mae:1.31970	valid-mae:1.33877
[23]	train-mae:1.31689	valid-mae:1.33621
[24]	train-mae:1.31404	val

[I 2022-05-14 20:53:51,901] Trial 15 finished with value: 1.2424869764337216 and parameters: {'eta': 0.11822802972781156}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.13012	valid-mae:2.11814
[1]	train-mae:1.94027	valid-mae:1.93020
[2]	train-mae:1.81109	valid-mae:1.80394
[3]	train-mae:1.71441	valid-mae:1.71087
[4]	train-mae:1.64165	valid-mae:1.63990
[5]	train-mae:1.58463	valid-mae:1.58417
[6]	train-mae:1.53968	valid-mae:1.54018
[7]	train-mae:1.50457	valid-mae:1.50643
[8]	train-mae:1.47548	valid-mae:1.47974
[9]	train-mae:1.45207	valid-mae:1.45624
[10]	train-mae:1.43168	valid-mae:1.43646
[11]	train-mae:1.41452	valid-mae:1.42012
[12]	train-mae:1.39867	valid-mae:1.40434
[13]	train-mae:1.38647	valid-mae:1.39306
[14]	train-mae:1.37597	valid-mae:1.38365
[15]	train-mae:1.36808	valid-mae:1.37635
[16]	train-mae:1.36059	valid-mae:1.36902
[17]	train-mae:1.35440	valid-mae:1.36342
[18]	train-mae:1.34921	valid-mae:1.35857
[19]	train-mae:1.34384	valid-mae:1.35405
[20]	train-mae:1.33972	valid-mae:1.35006
[21]	train-mae:1.33515	valid-mae:1.34592
[22]	train-mae:1.33149	valid-mae:1.34276
[23]	train-mae:1.32818	valid-mae:1.33953
[24]	train-mae:1.32500	val

[I 2022-05-14 20:56:08,188] Trial 16 finished with value: 1.2381578007096135 and parameters: {'eta': 0.10476838307699596}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.07609	valid-mae:2.07212
[1]	train-mae:1.79430	valid-mae:1.79426
[2]	train-mae:1.66247	valid-mae:1.66262
[3]	train-mae:1.57326	valid-mae:1.57697
[4]	train-mae:1.51275	valid-mae:1.51819
[5]	train-mae:1.47059	valid-mae:1.47725
[6]	train-mae:1.44060	valid-mae:1.44740
[7]	train-mae:1.41795	valid-mae:1.42353
[8]	train-mae:1.39937	valid-mae:1.40657
[9]	train-mae:1.38585	valid-mae:1.39349
[10]	train-mae:1.37269	valid-mae:1.38012
[11]	train-mae:1.36435	valid-mae:1.37201
[12]	train-mae:1.35749	valid-mae:1.36581
[13]	train-mae:1.35044	valid-mae:1.35937
[14]	train-mae:1.34387	valid-mae:1.35275
[15]	train-mae:1.33847	valid-mae:1.34806
[16]	train-mae:1.33374	valid-mae:1.34434
[17]	train-mae:1.32928	valid-mae:1.34099
[18]	train-mae:1.32546	valid-mae:1.33756
[19]	train-mae:1.32151	valid-mae:1.33451
[20]	train-mae:1.31831	valid-mae:1.33132
[21]	train-mae:1.31426	valid-mae:1.32839
[22]	train-mae:1.31056	valid-mae:1.32525
[23]	train-mae:1.30791	valid-mae:1.32315
[24]	train-mae:1.30412	val

[I 2022-05-14 20:58:20,614] Trial 17 finished with value: 1.2377360502912795 and parameters: {'eta': 0.1662025747600627}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.36227	valid-mae:2.35552
[1]	train-mae:2.24961	valid-mae:2.24374
[2]	train-mae:2.16418	valid-mae:2.15847
[3]	train-mae:2.09152	valid-mae:2.08625
[4]	train-mae:2.02527	valid-mae:2.02094
[5]	train-mae:1.96539	valid-mae:1.96186
[6]	train-mae:1.91311	valid-mae:1.91017
[7]	train-mae:1.86604	valid-mae:1.86352
[8]	train-mae:1.82399	valid-mae:1.82212
[9]	train-mae:1.78514	valid-mae:1.78377
[10]	train-mae:1.74956	valid-mae:1.74883
[11]	train-mae:1.71685	valid-mae:1.71768
[12]	train-mae:1.68770	valid-mae:1.68930
[13]	train-mae:1.66132	valid-mae:1.66387
[14]	train-mae:1.63658	valid-mae:1.64018
[15]	train-mae:1.61393	valid-mae:1.61854
[16]	train-mae:1.59301	valid-mae:1.59883
[17]	train-mae:1.57331	valid-mae:1.58000
[18]	train-mae:1.55410	valid-mae:1.56169
[19]	train-mae:1.53715	valid-mae:1.54508
[20]	train-mae:1.52127	valid-mae:1.52962
[21]	train-mae:1.50660	valid-mae:1.51513
[22]	train-mae:1.49306	valid-mae:1.50179
[23]	train-mae:1.48020	valid-mae:1.48941
[24]	train-mae:1.46874	val

[I 2022-05-14 21:00:40,820] Trial 18 finished with value: 1.2955088323653345 and parameters: {'eta': 0.035988574103694666}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.17434	valid-mae:2.16318
[1]	train-mae:2.01865	valid-mae:2.00818
[2]	train-mae:1.89723	valid-mae:1.88839
[3]	train-mae:1.80226	valid-mae:1.79449
[4]	train-mae:1.72691	valid-mae:1.72100
[5]	train-mae:1.66598	valid-mae:1.66144
[6]	train-mae:1.61505	valid-mae:1.61211
[7]	train-mae:1.57228	valid-mae:1.57118
[8]	train-mae:1.53669	valid-mae:1.53738
[9]	train-mae:1.50611	valid-mae:1.50723
[10]	train-mae:1.48037	valid-mae:1.48224
[11]	train-mae:1.45824	valid-mae:1.46083
[12]	train-mae:1.44000	valid-mae:1.44324
[13]	train-mae:1.42404	valid-mae:1.42803
[14]	train-mae:1.40968	valid-mae:1.41409
[15]	train-mae:1.39743	valid-mae:1.40244
[16]	train-mae:1.38650	valid-mae:1.39137
[17]	train-mae:1.37736	valid-mae:1.38305
[18]	train-mae:1.36879	valid-mae:1.37487
[19]	train-mae:1.36195	valid-mae:1.36752
[20]	train-mae:1.35624	valid-mae:1.36240
[21]	train-mae:1.35049	valid-mae:1.35679
[22]	train-mae:1.34517	valid-mae:1.35218
[23]	train-mae:1.34035	valid-mae:1.34791
[24]	train-mae:1.33568	val

[I 2022-05-14 21:02:57,069] Trial 19 finished with value: 1.252747468239225 and parameters: {'eta': 0.08080058885553983}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.09255	valid-mae:2.08525
[1]	train-mae:1.87146	valid-mae:1.86732
[2]	train-mae:1.71559	valid-mae:1.71527
[3]	train-mae:1.62703	valid-mae:1.63155
[4]	train-mae:1.55871	valid-mae:1.56703
[5]	train-mae:1.51280	valid-mae:1.52165
[6]	train-mae:1.47963	valid-mae:1.48926
[7]	train-mae:1.45329	valid-mae:1.46384
[8]	train-mae:1.42763	valid-mae:1.43891
[9]	train-mae:1.41305	valid-mae:1.42517
[10]	train-mae:1.39759	valid-mae:1.40921
[11]	train-mae:1.38491	valid-mae:1.39687
[12]	train-mae:1.37532	valid-mae:1.38883
[13]	train-mae:1.36776	valid-mae:1.38196
[14]	train-mae:1.35953	valid-mae:1.37474
[15]	train-mae:1.35340	valid-mae:1.36883
[16]	train-mae:1.34737	valid-mae:1.36358
[17]	train-mae:1.34207	valid-mae:1.35918
[18]	train-mae:1.33848	valid-mae:1.35691
[19]	train-mae:1.33406	valid-mae:1.35254
[20]	train-mae:1.32962	valid-mae:1.34926
[21]	train-mae:1.32631	valid-mae:1.34650
[22]	train-mae:1.32342	valid-mae:1.34285
[23]	train-mae:1.32082	valid-mae:1.34072
[24]	train-mae:1.31739	val

[I 2022-05-14 21:05:07,157] Trial 20 finished with value: 1.2475864355854485 and parameters: {'eta': 0.14258868431810234}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.06849	valid-mae:2.06703
[1]	train-mae:1.76007	valid-mae:1.75990
[2]	train-mae:1.63212	valid-mae:1.63594
[3]	train-mae:1.54727	valid-mae:1.55468
[4]	train-mae:1.49577	valid-mae:1.50241
[5]	train-mae:1.45634	valid-mae:1.46619
[6]	train-mae:1.42964	valid-mae:1.43886
[7]	train-mae:1.41135	valid-mae:1.41977
[8]	train-mae:1.39709	valid-mae:1.40439
[9]	train-mae:1.38564	valid-mae:1.39371
[10]	train-mae:1.37547	valid-mae:1.38603
[11]	train-mae:1.36787	valid-mae:1.38071
[12]	train-mae:1.36008	valid-mae:1.37356
[13]	train-mae:1.35361	valid-mae:1.36840
[14]	train-mae:1.34812	valid-mae:1.36410
[15]	train-mae:1.34244	valid-mae:1.35969
[16]	train-mae:1.33697	valid-mae:1.35499
[17]	train-mae:1.33269	valid-mae:1.35199
[18]	train-mae:1.32769	valid-mae:1.34801
[19]	train-mae:1.32490	valid-mae:1.34482
[20]	train-mae:1.32125	valid-mae:1.34178
[21]	train-mae:1.31750	valid-mae:1.33907
[22]	train-mae:1.31321	valid-mae:1.33560
[23]	train-mae:1.31040	valid-mae:1.33281
[24]	train-mae:1.30586	val

[I 2022-05-14 21:07:18,391] Trial 21 finished with value: 1.2407040046965034 and parameters: {'eta': 0.18803685012423432}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.08522	valid-mae:2.09145
[1]	train-mae:1.66832	valid-mae:1.67365
[2]	train-mae:1.59286	valid-mae:1.60625
[3]	train-mae:1.54673	valid-mae:1.56006
[4]	train-mae:1.50987	valid-mae:1.52606
[5]	train-mae:1.48763	valid-mae:1.50179
[6]	train-mae:1.47133	valid-mae:1.48641
[7]	train-mae:1.45865	valid-mae:1.47653
[8]	train-mae:1.44656	valid-mae:1.46333
[9]	train-mae:1.43555	valid-mae:1.45238
[10]	train-mae:1.42740	valid-mae:1.44422
[11]	train-mae:1.42068	valid-mae:1.43876
[12]	train-mae:1.40156	valid-mae:1.41755
[13]	train-mae:1.39457	valid-mae:1.41188
[14]	train-mae:1.38776	valid-mae:1.40671
[15]	train-mae:1.38069	valid-mae:1.40078
[16]	train-mae:1.37388	valid-mae:1.39566
[17]	train-mae:1.36768	valid-mae:1.39076
[18]	train-mae:1.36239	valid-mae:1.38666
[19]	train-mae:1.35579	valid-mae:1.38245
[20]	train-mae:1.35141	valid-mae:1.37952
[21]	train-mae:1.34648	valid-mae:1.37426
[22]	train-mae:1.34231	valid-mae:1.36990
[23]	train-mae:1.33647	valid-mae:1.36232
[24]	train-mae:1.33233	val

[I 2022-05-14 21:09:22,276] Trial 22 finished with value: 1.2667110244156405 and parameters: {'eta': 0.29707578191848716}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.08657	valid-mae:2.08044
[1]	train-mae:1.81910	valid-mae:1.81700
[2]	train-mae:1.68823	valid-mae:1.69017
[3]	train-mae:1.59815	valid-mae:1.60332
[4]	train-mae:1.53476	valid-mae:1.54036
[5]	train-mae:1.49001	valid-mae:1.49580
[6]	train-mae:1.45733	valid-mae:1.46333
[7]	train-mae:1.43053	valid-mae:1.43558
[8]	train-mae:1.41093	valid-mae:1.41489
[9]	train-mae:1.39630	valid-mae:1.40151
[10]	train-mae:1.38164	valid-mae:1.38722
[11]	train-mae:1.37241	valid-mae:1.37875
[12]	train-mae:1.36386	valid-mae:1.37101
[13]	train-mae:1.35694	valid-mae:1.36572
[14]	train-mae:1.35043	valid-mae:1.36170
[15]	train-mae:1.34556	valid-mae:1.35700
[16]	train-mae:1.34006	valid-mae:1.35284
[17]	train-mae:1.33537	valid-mae:1.34851
[18]	train-mae:1.33163	valid-mae:1.34646
[19]	train-mae:1.32827	valid-mae:1.34361
[20]	train-mae:1.32475	valid-mae:1.34122
[21]	train-mae:1.32065	valid-mae:1.33748
[22]	train-mae:1.31715	valid-mae:1.33543
[23]	train-mae:1.31318	valid-mae:1.33183
[24]	train-mae:1.30957	val

[I 2022-05-14 21:11:34,649] Trial 23 finished with value: 1.2411060521469783 and parameters: {'eta': 0.15042475869841884}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.06878	valid-mae:2.06716
[1]	train-mae:1.76282	valid-mae:1.76258
[2]	train-mae:1.63496	valid-mae:1.63939
[3]	train-mae:1.54968	valid-mae:1.55757
[4]	train-mae:1.49423	valid-mae:1.50614
[5]	train-mae:1.45736	valid-mae:1.46802
[6]	train-mae:1.43215	valid-mae:1.44301
[7]	train-mae:1.41228	valid-mae:1.42072
[8]	train-mae:1.39830	valid-mae:1.40679
[9]	train-mae:1.38678	valid-mae:1.39659
[10]	train-mae:1.37746	valid-mae:1.38913
[11]	train-mae:1.36889	valid-mae:1.38168
[12]	train-mae:1.36181	valid-mae:1.37579
[13]	train-mae:1.35494	valid-mae:1.37142
[14]	train-mae:1.34874	valid-mae:1.36655
[15]	train-mae:1.34450	valid-mae:1.36261
[16]	train-mae:1.34021	valid-mae:1.35866
[17]	train-mae:1.33531	valid-mae:1.35561
[18]	train-mae:1.33180	valid-mae:1.35221
[19]	train-mae:1.32694	valid-mae:1.34936
[20]	train-mae:1.32341	valid-mae:1.34605
[21]	train-mae:1.32002	valid-mae:1.34325
[22]	train-mae:1.31709	valid-mae:1.34079
[23]	train-mae:1.31313	valid-mae:1.33743
[24]	train-mae:1.30887	val

[I 2022-05-14 21:13:41,867] Trial 24 finished with value: 1.2481263409824919 and parameters: {'eta': 0.18635432446452413}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.15605	valid-mae:2.14396
[1]	train-mae:1.99199	valid-mae:1.98083
[2]	train-mae:1.86728	valid-mae:1.85851
[3]	train-mae:1.77163	valid-mae:1.76418
[4]	train-mae:1.69850	valid-mae:1.69301
[5]	train-mae:1.64025	valid-mae:1.63558
[6]	train-mae:1.59125	valid-mae:1.58786
[7]	train-mae:1.55075	valid-mae:1.54969
[8]	train-mae:1.51667	valid-mae:1.51603
[9]	train-mae:1.48764	valid-mae:1.48861
[10]	train-mae:1.46276	valid-mae:1.46595
[11]	train-mae:1.44280	valid-mae:1.44706
[12]	train-mae:1.42591	valid-mae:1.43104
[13]	train-mae:1.41108	valid-mae:1.41745
[14]	train-mae:1.39849	valid-mae:1.40445
[15]	train-mae:1.38801	valid-mae:1.39417
[16]	train-mae:1.37855	valid-mae:1.38532
[17]	train-mae:1.37002	valid-mae:1.37745
[18]	train-mae:1.36347	valid-mae:1.37146
[19]	train-mae:1.35652	valid-mae:1.36600
[20]	train-mae:1.35135	valid-mae:1.36165
[21]	train-mae:1.34619	valid-mae:1.35707
[22]	train-mae:1.34186	valid-mae:1.35304
[23]	train-mae:1.33788	valid-mae:1.34963
[24]	train-mae:1.33405	val

[I 2022-05-14 21:15:58,767] Trial 25 finished with value: 1.2499675682923586 and parameters: {'eta': 0.08855139577249355}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.32957	valid-mae:2.32255
[1]	train-mae:2.21146	valid-mae:2.20440
[2]	train-mae:2.11777	valid-mae:2.11160
[3]	train-mae:2.03863	valid-mae:2.03356
[4]	train-mae:1.96686	valid-mae:1.96314
[5]	train-mae:1.90431	valid-mae:1.90146
[6]	train-mae:1.85026	valid-mae:1.84870
[7]	train-mae:1.80268	valid-mae:1.80170
[8]	train-mae:1.75973	valid-mae:1.76037
[9]	train-mae:1.72086	valid-mae:1.72221
[10]	train-mae:1.68609	valid-mae:1.68829
[11]	train-mae:1.65497	valid-mae:1.65844
[12]	train-mae:1.62603	valid-mae:1.63090
[13]	train-mae:1.60051	valid-mae:1.60549
[14]	train-mae:1.57685	valid-mae:1.58256
[15]	train-mae:1.55587	valid-mae:1.56238
[16]	train-mae:1.53599	valid-mae:1.54304
[17]	train-mae:1.51797	valid-mae:1.52548
[18]	train-mae:1.50181	valid-mae:1.50909
[19]	train-mae:1.48702	valid-mae:1.49404
[20]	train-mae:1.47335	valid-mae:1.48026
[21]	train-mae:1.46075	valid-mae:1.46781
[22]	train-mae:1.44920	valid-mae:1.45638
[23]	train-mae:1.43802	valid-mae:1.44536
[24]	train-mae:1.42772	val

[I 2022-05-14 21:18:18,057] Trial 26 finished with value: 1.2872195867984781 and parameters: {'eta': 0.04241065290962887}. Best is trial 10 with value: 1.234455828166958.


[0]	train-mae:2.09964	valid-mae:2.09125
[1]	train-mae:1.90390	valid-mae:1.90533
[2]	train-mae:1.73416	valid-mae:1.72737
[3]	train-mae:1.63289	valid-mae:1.62971
[4]	train-mae:1.56062	valid-mae:1.56143
[5]	train-mae:1.51209	valid-mae:1.51506
[6]	train-mae:1.47620	valid-mae:1.47942
[7]	train-mae:1.44729	valid-mae:1.45106
[8]	train-mae:1.42366	valid-mae:1.42802
[9]	train-mae:1.40627	valid-mae:1.41151
[10]	train-mae:1.39235	valid-mae:1.39812
[11]	train-mae:1.38005	valid-mae:1.38640
[12]	train-mae:1.37082	valid-mae:1.37785
[13]	train-mae:1.36234	valid-mae:1.37056
[14]	train-mae:1.35440	valid-mae:1.36363
[15]	train-mae:1.34760	valid-mae:1.35835
[16]	train-mae:1.34065	valid-mae:1.35202
[17]	train-mae:1.33546	valid-mae:1.34808
[18]	train-mae:1.33133	valid-mae:1.34435
[19]	train-mae:1.32758	valid-mae:1.34079
[20]	train-mae:1.32424	valid-mae:1.33870
[21]	train-mae:1.31992	valid-mae:1.33515
[22]	train-mae:1.31558	valid-mae:1.33308
[23]	train-mae:1.31188	valid-mae:1.33053
[24]	train-mae:1.30803	val

KeyboardInterrupt: 

Number of finished trials:  3  
Best trial:  
  Value: 1.2337084043245454  
  Params:   
    eta: 0.13509680253299897  
    
Trial 10 finished with value: 1.234455828166958 and parameters: {'eta': 0.15406470353245602}.

Use tuned parameters and train xgboost on cpu on the whole dataset

In [71]:
import multiprocessing

In [32]:
xgb_reg = xgb.XGBRegressor(
    n_jobs=multiprocessing.cpu_count() // 2,
    num_round=XGB_TRAIN_ROUNDS,
    max_leaves=XGB_NUM_LEAVES,
    eval_metric="mae",
    objective="reg:pseudohubererror",
)
bst_xgb = xgb_reg.fit(X_train_sparse, y_train)
xgb_test_pred = bst_xgb.predict(X_test_sparse.astype(np.float32))

[19:49:56] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.6.0/src/learner.cc:627: 
Parameters: { "num_round" } might not be used.

  This could be a false alarm, with some parameters getting used by language bindings but
  then being mistakenly passed down to XGBoost core, or some parameter actually being used
  but getting flagged wrongly here. Please open an issue if you find any such cases.




C:\Users\ryblo\AppData\Local\Temp\ipykernel_17936\2221456.py:9: RuntimeWarning: overflow encountered in expm1
  xgb_test_pred = np.expm1(bst_xgb.predict(X_test_sparse.astype(np.float32)))


## Training with all data

**Train the same Ridge with all available data, make predictions for the test set and form a submission file.**

In [48]:
mix_pred = XGB_WEIGHT * xgb_test_pred + RIDGE_WEIGHT * ridge_test_pred
mix_test_pred_modif = ridge_test_pred + MEAN_TEST_TARGET - y_train.mean()

NameError: name 'XGB_WEIGHT' is not defined

In [50]:
def write_submission_file(
    prediction,
    filename,
    path_to_sample=os.path.join(PATH_TO_DATA, "sample_submission.csv"),
):
    submission = pd.read_csv(path_to_sample, index_col="id")

    submission["log_recommends"] = prediction
    submission.to_csv(filename)

In [70]:
write_submission_file(mix_test_pred_modif, os.path.join(PATH_TO_DATA, "assignment6_medium_submission.csv"))

**Now's the time for dirty Kaggle hacks. Form a submission file with all zeros. Make a submission. What do you get if you think about it? How is it going to help you with modifying your predictions?**

In [18]:
write_submission_file(
    np.zeros_like(ridge_test_pred),
    os.path.join(PATH_TO_DATA, "medium_all_zeros_submission.csv"),
)

**Modify predictions in an appropriate way (based on your all-zero submission) and make a new submission.**

In [19]:
ridge_test_pred_modif = ridge_test_pred
# You code here (read-only in a JupyterBook, pls run jupyter-notebook to edit)

In [20]:
write_submission_file(
    ridge_test_pred_modif,
    os.path.join(PATH_TO_DATA, "assignment6_medium_submission_with_hack.csv"),
)

That's it for the assignment. In case you'd like to try some more ideas for improvement:

- Engineer good features, this is the key to success. Some simple features will be based on publication time, authors, content length and so on
- You may not ignore HTML and extract some features from there
- You'd better experiment with your validation scheme. You should see a correlation between your local improvements and LB score
- Try TF-IDF, ngrams, Word2Vec and GloVe embeddings
- Try various NLP techniques like stemming and lemmatization
- Tune hyperparameters. In our example, we've left only 50k features and used C=1 as a regularization parameter, this can be changed
- SGD and Vowpal Wabbit will train much faster
- Play around with blending and/or stacking. An intro is given in [this Kernel](https://www.kaggle.com/kashnitsky/ridge-and-lightgbm-simple-blending) by @yorko 
- And neural nets of course. We don't cover them in this course byt still transformer-based architectures will likely perform well in such types of tasks